<!DOCTYPE html>
<html lang="en">
<head>
    <!-- Google tag (gtag.js) -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=G-5DYZK6SY5M"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'G-5DYZK6SY5M');
    </script>
    <!-- Required meta tags (add the responsive viewport meta tag) -->
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="description"
        content="single cell lncRNA database web application">
    <title>Single Cell LncRNA Atlas</title>
    
    <!--Bootstrap CSS library-->
    <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/css/bootstrap.min.css" integrity="sha384-JcKb8q3iqJ61gNV9KGb8thSsNjpSL0n8PARn9HuZOnIxN0hoP+VmmDGMN5t9UJ0Z" crossorigin="anonymous">
    <!--Google Icon CSS library-->
    <link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">
    <!-- JS -->
    <!--For Using utility functions-->
    <script src="https://cdn.jsdelivr.net/npm/lodash@4.17.19/lodash.min.js"></script>
    <script> lodash = _.noConflict(); // load script of lodash </script> 
    <!--Plotly-->
    <script src="https://cdn.plot.ly/plotly-1.57.1.min.js"></script>
    <!--For loading large csv files-->
    <script src="https://cdnjs.cloudflare.com/ajax/libs/PapaParse/5.2.0/papaparse.min.js"></script>
    <!--Python's Numpy-like Library, NumJS-->
    <script src="https://cdn.jsdelivr.net/gh/nicolaspanel/numjs@0.15.1/dist/numjs.min.js"></script>
    <!--D3 to draw tables-->
    <script src="https://cdnjs.cloudflare.com/ajax/libs/d3/3.5.5/d3.min.js"></script>
    <!--chroma to draw colormaps-->
    <script src="https://cdnjs.cloudflare.com/ajax/libs/chroma-js/2.1.0/chroma.min.js"></script>
    <!--hotkeys to add keyboard shortcuts-->
    <script src="https://cdn.jsdelivr.net/npm/hotkeys-js@3.8.1/dist/hotkeys.min.js" integrity="sha256-4nmnbhmVpr2Rg64+7ysAr93+vQrfc0UvHWwm4W6uXJM=" crossorigin="anonymous"></script>
    
    <!--Datatable for efficient visualization of tabular data-->
    <!--full: jquery.dataTables.js, jquery.dataTables.css-->
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js" crossorigin="anonymous"></script>
    <link rel="stylesheet" type="text/css" href="https://cdn.datatables.net/1.10.21/css/jquery.dataTables.min.css">
    <script type="text/javascript" charset="utf8" src="https://cdn.datatables.net/1.10.21/js/jquery.dataTables.min.js" crossorigin="anonymous"></script>
    <!-- jQuery first, then Popper.js, then Bootstrap JS -->
    <script src="https://cdn.jsdelivr.net/npm/popper.js@1.16.1/dist/umd/popper.min.js" integrity="sha384-9/reFTGAW83EW2RDu2S0VKaIzap3H66lZH81PoYlFhbGU+6BZp6G7niu735Sk7lN" crossorigin="anonymous"></script>
    <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/js/bootstrap.min.js" integrity="sha384-B4gt1jrGC7Jh4AgTPSdUtOBvfO8shuf57BaghqFfPlYxofvL8/KUEfYiJOMMV+rV" crossorigin="anonymous"></script>
    <!-- dataframe.js for useful dataframe functions -->
    <script src="https://cdn.jsdelivr.net/npm/dataframe-js@1.4.3/dist/dataframe.min.js" integrity="sha256-ncD0rnpkULRQhB+hFzhFCQ4KBu+NKVYEQHyhxsUA29Q=" crossorigin="anonymous"></script>
    <!-- FileSaver.js to save js object as local files -->
    <script src="https://cdn.jsdelivr.net/npm/filesaver.js@1.3.4/FileSaver.min.js" integrity="sha256-5IjlS/EkXaEXRrVOMusSOlmQsyFy3DvCD5QC5mJZubU=" crossorigin="anonymous"></script>
    <!-- Html2Canvas.js to save current view as a PNG file -->
    <script src="https://cdn.jsdelivr.net/npm/html2canvas@1.0.0-rc.7/dist/html2canvas.min.js" integrity="sha256-Ax1aqtvxWBY0xWND+tPZVva/VQZy9t1Ce17ZJO+NTRc=" crossorigin="anonymous"></script>
    <!-- Pako.js to parse base64 gzipped text file downloaded from remote location -->
    <script src="https://cdn.jsdelivr.net/npm/pako@1.0.11/dist/pako.min.js" integrity="sha256-Kc+gVCuYZLZkDP3MjxWxhNtkMbUy2ycCo86X5fKn/Bw=" crossorigin="anonymous"></script>
    <!-- Bootstrap-Toggle.js for responsive toggle button -->
    <link href="https://cdn.jsdelivr.net/gh/gitbrent/bootstrap4-toggle@3.6.1/css/bootstrap4-toggle.min.css" rel="stylesheet">
    <script src="https://cdn.jsdelivr.net/gh/gitbrent/bootstrap4-toggle@3.6.1/js/bootstrap4-toggle.min.js"></script>
<!--   Tagify for search bar interaction   -->
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@yaireo/tagify@3.21.5/dist/tagify.css" integrity="sha256-JQCBuvZzuXrKFdHRTVW23P/DqnFsfvbI3kraGJhPyNA=" crossorigin="anonymous">
    <script src="https://cdn.jsdelivr.net/npm/@yaireo/tagify@3.21.5/dist/tagify.min.js" integrity="sha256-6QmUlbXKkm+RVqvqtIzhhDQ3cqRkuREeUhNJDnL2ujM=" crossorigin="anonymous"></script>
<!--   vis-network.js for graph visualization   -->
     <script type="text/javascript" src="https://unpkg.com/vis-network/standalone/umd/vis-network.min.js"></script>
    
<!--  Font Awesome    -->
    <script src="https://kit.fontawesome.com/b773d81884.js" crossorigin="anonymous"></script>
<!--     <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.8.2/css/all.css" integrity="sha384-oS3vJWv+0UjzBfQzYUhtDYW+Pj2yciDJxpsK1OYPAYjqT085Qq/1cq5FLXAZQ7Ay" crossorigin="anonymous"> -->
    <link rel="stylesheet" href="css/github__pro_sidebar-jquery.mCustomScrollbar.min.css">
    <link rel="stylesheet" href="css/github__pro_sidebar-main.css">
    <link rel="stylesheet" href="css/github__pro_sidebar-sidebar-themes.css">
    
<!--  for responsive sidebar (from https://github.com/azouaoui-med/pro-sidebar-template)    -->
    
    <style>
      * { box-sizing: border-box; }
        /* Create two equal columns that floats next to each other */
        .column {
          float: left;
          width: 50%;
          padding: 10px;
          height: 450px; /* Should be removed. Only for demonstration */
          position:relative;
        }
        /* Clear floats after the columns */
        .responsive_row_with_two_columns:after {
          float:left;
          content: "";
          display: table;
          clear: both;
        }
        /* Responsive layout - makes the two columns stack on top of each other instead of next to each other */
        @media screen and (max-width: 700px) {
          .column {
            width: 100%;
          }
        }
        .deconstructed_pancake_box { /* example in https://web.dev/one-line-layouts/ */
            flex: 1 1 450px;
            margin: 5px;
        }
        .deconstructed_pancake_parent {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
        }
        .container_molstar {
          height: 600px; /* Should be modified */
          position:relative;
        }
        #container_d3_table_metrics{
          display: grid;
          grid-template-rows: auto 1fr auto;
          position:relative;
        }
        #plotly_graph_molecular_mimicry_map{
            float:left;
            width:100%;
            padding: 5px;
            height:575px;
            position:relative;
            overflow:hidden;
        }
        .graph_for_selection{
            float:left;
            width:100%;
            height:350px;
            position:relative;
        }
/*    custom bootstrap button color   */
        .bg-custom-navbar{  /* custom navbar color and gradient for secondary navbar (color representing lupus awareness color) */
            background-image: linear-gradient( 20deg, #004d4d 0%, #00e0e0 100%);
        }
        .bg-custom-navbar-secondary{  /* custom navbar color and gradient (color representing lupus awareness color) */
            background-image: linear-gradient( 15deg, #46137a 0%, #f08dee 100%);
        }
        .bg-custom-navbar-hidden{  /* custom navbar color and gradient (color representing lupus awareness color) */
            border-left: 6px solid #ccc;
            border-bottom: 3px solid #ccc;
            color: #000;
            border-color: #0a5d63;
            background-color: #f8f7fa;
/*             '#fce6ff; #dfdae6; #f3f0f7; */
        }
        .bg-custom-navbar-top{  /* custom navbar color and gradient (color representing lupus awareness color) */
            background-image: linear-gradient( 0deg, #ababab 0%, #f7f7f7 100%);
        }
        /* CSS for D3 Table */
        table { border-collapse: collapse; color: #333; background-color: #F7F6F3; margin: auto; } /* place the table in the middle of div */
        table thead { font-weight: bold; background-color: #CCC; cursor: default; text-align: center; }
        table tbody tr:hover { background-color: #FFC; }
        td { border: solid 1px #CCC; padding: 0 1ex; }
        .table_even { color: #284775; background-color: White; }
        .table_left { text-align: left; }
        .table_center { text-align: center; }
        .table_right { text-align: right; }
        .table_add { color: green; }
        .table_minus { color: red; }

/*     for tidy organization of buttons     */
        .btn { 
            margin-bottom: 0px !important;
        }
        .helpIcon
        {
          cursor: pointer;
          padding-top:12px;
        }
        i.icon-white {
            color: white;
            display:inline-flex;
            vertical-align:middle;
        }
        i.icon-black {
            color: black;
            display:inline-flex;
            vertical-align:middle;
        }
        /* for labeled buttons */
        .btn-label {position: relative;left: -12px;display: inline-block;padding: 6px 12px;background: rgba(0,0,0,0.1);border-radius: 3px 0 0 3px;}
        .btn-labeled {padding-top: 0;padding-bottom: 0;}
        .btn { margin-bottom:10px; }
        /* for  div. displayed during loading (https://stackoverflow.com/questions/54439548/display-loading-symbol-while-waiting-for-a-result-with-plot-ly-dash) */
        @keyframes fadein {
            0% {
                opacity: 0;
            }
            100% {
                opacity: 0.5;
            }
        }
        ._plotly-loading-callback {
          font-family: sans-serif;
          padding-top: 50px;
          color: rgb(90, 90, 90);

          /* The banner */
          position: fixed;
          top: 0;
          left: 0;
          width: 100%;
          height: 100%;
          text-align: center;
          cursor: progress;

          opacity: 0;
          background-color: rgb(250, 250, 250);
          /* Delay animation by 1s to prevent flashing 
             and smoothly transition the animation over 0.5s 
           */
          -moz-animation: fadein 0.5s ease-in 1s forwards; /* Firefox */
          -webkit-animation: fadein 0.5s ease-in 1s forwards; /* Safari and Chrome */
          -o-animation: fadein 0.5s ease-in 1s forwards; /* Opera */
            animation: fadein 0.5s ease-in 1s forwards;
        }
        /* Add an extra .carousel parent class to increase specificity
           avoiding the use of !important flag. */
        .carousel-caption {
            color: black;
        }
        .carousel .carousel-indicators li {
            background-color: #fff;
            background-color: rgba(70, 70, 70, 0.25);
        }

        .carousel .carousel-indicators .active {
            background-color: #444;
        }
        .carousel .carousel-indicators li {
            background-color: #fff;
            background-color: rgba(70, 70, 70, 0.25);
        }

        .carousel .carousel-indicators .active {
            background-color: #444;
        }
        .carousel-control-next,
        .carousel-control-prev {
            filter: invert(100%);
        }
        /* for loading SVG images */
/*         .loader{
          margin: 0 0 2em;
          height: 100px;
          width: 100%;
          text-align: center;
          padding: 1em;
          margin: 0 auto 1em;
          display: inline-block;
          vertical-align: top;
        } */
/*  https://gitbrent.github.io/bootstrap4-toggle/ custom settings  */
          .toggle.ios, .toggle-on.ios, .toggle-off.ios { border-radius: 20rem; height: 35px !important; }
          .toggle.ios .toggle-handle { border-radius: 20rem; height: 35px !important; }
/* twitter typeahead css */
        .tt-query {
          -webkit-box-shadow: inset 0 1px 1px rgba(0, 0, 0, 0.075);
             -moz-box-shadow: inset 0 1px 1px rgba(0, 0, 0, 0.075);
                  box-shadow: inset 0 1px 1px rgba(0, 0, 0, 0.075);
        }
        .tt-hint {
          color: #999
        }
        .tt-menu {    /* used to be tt-dropdown-menu in older versions */
          width: 422px;
          margin-top: 4px;
          padding: 4px 0;
          background-color: #fff;
          border: 1px solid #ccc;
          border: 1px solid rgba(0, 0, 0, 0.2);
          -webkit-border-radius: 4px;
             -moz-border-radius: 4px;
                  border-radius: 4px;
          -webkit-box-shadow: 0 5px 10px rgba(0,0,0,.2);
             -moz-box-shadow: 0 5px 10px rgba(0,0,0,.2);
                  box-shadow: 0 5px 10px rgba(0,0,0,.2);
        }
        .tt-suggestion {
          padding: 3px 20px;
          line-height: 24px;
        }
        .tt-suggestion.tt-cursor,.tt-suggestion:hover {
          color: #fff;
          background-color: #0097cf;

        }
        .tt-suggestion p {
          margin: 0;
        }
/*     deconstructed pancake     */
        .deconstructed_pancake__parent {
            display: flex;
            flex-wrap: wrap;
            justify-content: center;
        }
        .deconstructed_pancake__box {
            flex: 1 1 750px;
            margin: 10px;
        }
        
/* sidebar button (for small window) */
        #button_open_sidebar { /* CSS for Scroll Back To Top Button */
          display: none; /* Hidden by default */
          position: fixed; /* Fixed/sticky position */
          top: 25px; /* Place the button at the bottom of the page */
          right: 30px; /* Place the button 30px from the right */
          z-index: 9999; /* Make sure it does not overlap */
          outline: none; /* Remove outline */
          border-width: 1px;
          border-color: #0a5e5d;
          background-color: #10918f; /* Set a background color */
          color: white; /* Text color */
          cursor: pointer; /* Add a mouse pointer on hover */
          padding: 15px; /* Some padding */
          border-radius: 50%; /* Rounded corners */
          font-size: 18px; /* Increase font size */
        }
        #button_open_sidebar:hover {
          background-color: #555 ; /* Indigo Add a dark-grey background on hover */
        }
        
/*     custom css on Pro Sidebar (https://github.com/azouaoui-med/pro-sidebar-template)  */
        
#mynetwork {
  width: 600px;
  height: 400px;
  border: 1px solid lightgray;
}
body {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen",
    "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue",
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}
code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, "Courier New",
    monospace;
}
    </style>
    <!-- icon of the document -->
    <link rel="icon" href="data:image/svg+xml,%3Csvg%20xmlns='http://www.w3.org/2000/svg'%20viewBox='0%200%2020%2020'%3E%3Ctext%20x='0'%20y='16'%20style='filter:%20invert(100%);'%3E🦠%3C/text%3E%3C/svg%3E" type="image/svg+xml" />
</head>
    
<body>
    
<!-- where captured current view is stored as a data URL  -->
<a id="anchor-current-view" href="#" hidden></a>
    <div id='container_main_sidebar' class="page-wrapper light-theme border-radius-on sidebar-bg bg1 toggled">
        <nav id="sidebar" class="sidebar-wrapper">
            <div class="sidebar-content">
                <!-- sidebar-brand  -->
<!--                 <div class="sidebar-item sidebar-brand">
                    <a href="#">Single Cell lncRNA Atlas</a>
                </div> -->
                <!-- sidebar-header  -->
<!--                 <div class="sidebar-item sidebar-header d-flex flex-nowrap">
                </div> -->
                <!-- sidebar-search  -->
<!--                 <div class="sidebar-item sidebar-search">
                    <div>
                        <div class="input-group">
                            <input name='tags' id='searchbox_cell_cluster' placeholder='Search Cell Cluster' value='' data-blacklist='.NET,PHP'>
                            <div class="input-group-append">
                                <span class="input-group-text">
                                    <i class="fa fa-search" aria-hidden="true"></i>
                                </span>
                            </div>
                        </div>
                    </div>
                </div> -->
                <!-- sidebar-menu  -->
                <div class=" sidebar-item sidebar-menu">
                    <ul>
                        <li class="header-menu">
                            <span>UMAP</span>
                        </li>
                        <li class="sidebar-dropdown">
                            <a href="#">
                                <i class="fas fa-lungs"></i>
                                <span class="menu-text">Aging/Young Kidney</span>
                            </a>
                            <div class="sidebar-submenu">
                                <ul>
                                    <li>
                                        <a href="#" onclick="Change_UMAP_Coordinate( 'Combined.Aging_Kidney' )">All<span class="badge badge-pill badge-primary">20261 cells</span></a>
                                    </li>
                                    <li>
                                    <li>
                                        <a href="#" onclick="Change_UMAP_Coordinate( 'Glom (Old)' )">Glomerulus (Old)<span class="badge badge-pill badge-secondary">6505 cells</span></a>
                                    </li>
                                    <li>
                                        <a href="#" onclick="Change_UMAP_Coordinate( 'Glom (Young)' )">Glomerulus (Young)<span class="badge badge-pill badge-secondary">7658 cells</span></a>
                                    </li>
                                    <li>
                                        <a href="#" onclick="Change_UMAP_Coordinate( 'Tubule (Old)' )">Tubule (Old)<span class="badge badge-pill badge-secondary">2772 cells</span></a>
                                    </li>
                                    <li>
                                        <a href="#" onclick="Change_UMAP_Coordinate( 'Tubule (Young)' )">Tubule (Young)<span class="badge badge-pill badge-secondary">3326 cells</span></a>
                                    </li>
                                </ul>
                            </div>
                        </li>
                        
                        <li class="sidebar-dropdown">
                            <a href="#">
                                <i class="fas fa-heart"></i>
                                <span class="menu-text">Diverse Mouse Tissues</span>
                            </a>
                            <div class="sidebar-submenu">
                                <ul>
                                    <li>
                                        <a href="#" onclick="Change_UMAP_Coordinate( 'Combined.mTissues6' )">All<span class="badge badge-pill badge-primary">22244 cells</span></a>
                                    </li>
                                    <li>
                                    <li>
                                        <a href="#" onclick="Change_UMAP_Coordinate( 'Heart' )">Heart<span class="badge badge-pill badge-secondary">5979 cells</span></a>
                                    </li>
                                    <li>
                                        <a href="#" onclick="Change_UMAP_Coordinate( 'Intestine' )">Intestine<span class="badge badge-pill badge-secondary">2330 cells</span></a>
                                    </li>
                                    <li>
                                        <a href="#" onclick="Change_UMAP_Coordinate( 'Kidney' )">Kidney<span class="badge badge-pill badge-secondary">2455 cells</span></a>
                                    </li>
                                    <li>
                                        <a href="#" onclick="Change_UMAP_Coordinate( 'Lung' )">Lung<span class="badge badge-pill badge-secondary">4375 cells</span></a>
                                    </li>
                                    <li>
                                        <a href="#" onclick="Change_UMAP_Coordinate( 'Liver' )">Liver<span class="badge badge-pill badge-secondary">4191 cells</span></a>
                                    </li>
                                    <li>
                                        <a href="#" onclick="Change_UMAP_Coordinate( 'Thymus' )">Thymus<span class="badge badge-pill badge-secondary">2914 cells</span></a>
                                    </li>
                                </ul>
                            </div>
                        </li>
                        
                        <li class="header-menu">
                            <span>Data Table</span>
                        </li>
                        <li>
                            <a href="#">
                                <i class="fas fa-viruses"></i>
                                <span class="menu-text" data-toggle="collapse" data-target="#container_panel_datatable_cell_marker" aria-expanded="false" aria-controls="container_panel_datatable_cell_marker">Cell Type Markers</span>
                            </a>
                        </li>
                        <li>
                            <a href="#">
                                <i class="fas fa-chart-line"></i>
                                <span class="menu-text" data-toggle="collapse" data-target="#container_panel_datatable_correl" aria-expanded="false" aria-controls="container_panel_datatable_correl">Co-expression</span>
                            </a>
                        </li>
<!--                         <li class="header-menu">
                            <span>Extra</span>
                        </li>
                        <li>
                            <a href="#">
                                <i class="fa fa-book"></i>
                                <span class="menu-text">Documentation</span>
                                <span class="badge badge-pill badge-primary">Beta</span>
                            </a>
                        </li> -->
                        <li class="header-menu">
                            <span>Pin sidebar</span>
                        </li>
                        <li>
                            <a id="pin-sidebar" class="btn btn-light" href="#">
                                <span><i class="fas fa-thumbtack"></i></span>
                            </a>
                        </li>
                    </ul>
                </div>
                <!-- sidebar-menu  -->
            </div>
            <!-- sidebar-footer  -->
<!--             <div class="sidebar-footer">
                <div class="dropdown">

                    <a href="#" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                        <i class="fa fa-bell"></i>
                        <span class="badge badge-pill badge-warning notification">3</span>
                    </a>
                    <div class="dropdown-menu notifications" aria-labelledby="dropdownMenuMessage">
                        <div class="notifications-header">
                            <i class="fa fa-bell"></i>
                            Notifications
                        </div>
                        <div class="dropdown-divider"></div>
                        <a class="dropdown-item" href="#">
                            <div class="notification-content">
                                <div class="icon">
                                    <i class="fas fa-check text-success border border-success"></i>
                                </div>
                                <div class="content">
                                    <div class="notification-detail">Lorem ipsum dolor sit amet consectetur adipisicing
                                        elit. In totam explicabo</div>
                                    <div class="notification-time">
                                        6 minutes ago
                                    </div>
                                </div>
                            </div>
                        </a>
                        <a class="dropdown-item" href="#">
                            <div class="notification-content">
                                <div class="icon">
                                    <i class="fas fa-exclamation text-info border border-info"></i>
                                </div>
                                <div class="content">
                                    <div class="notification-detail">Lorem ipsum dolor sit amet consectetur adipisicing
                                        elit. In totam explicabo</div>
                                    <div class="notification-time">
                                        Today
                                    </div>
                                </div>
                            </div>
                        </a>
                        <a class="dropdown-item" href="#">
                            <div class="notification-content">
                                <div class="icon">
                                    <i class="fas fa-exclamation-triangle text-warning border border-warning"></i>
                                </div>
                                <div class="content">
                                    <div class="notification-detail">Lorem ipsum dolor sit amet consectetur adipisicing
                                        elit. In totam explicabo</div>
                                    <div class="notification-time">
                                        Yesterday
                                    </div>
                                </div>
                            </div>
                        </a>
                        <div class="dropdown-divider"></div>
                        <a class="dropdown-item text-center" href="#">View all notifications</a>
                    </div>
                </div>
                <div class="dropdown">
                    <a href="#" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                        <i class="fa fa-envelope"></i>
                        <span class="badge badge-pill badge-success notification">7</span>
                    </a>
                    <div class="dropdown-menu messages" aria-labelledby="dropdownMenuMessage">
                        <div class="messages-header">
                            <i class="fa fa-envelope"></i>
                            Messages
                        </div>
                        <div class="dropdown-divider"></div>
                        <a class="dropdown-item" href="#">
                            <div class="message-content">
                                <div class="pic">
                                    <img src="img/user.jpg" alt="">
                                </div>
                                <div class="content">
                                    <div class="message-title">
                                        <strong> Jhon doe</strong>
                                    </div>
                                    <div class="message-detail">Lorem ipsum dolor sit amet consectetur adipisicing
                                        elit. In totam explicabo</div>
                                </div>
                            </div>

                        </a>
                        <a class="dropdown-item" href="#">
                            <div class="message-content">
                                <div class="pic">
                                    <img src="img/user.jpg" alt="">
                                </div>
                                <div class="content">
                                    <div class="message-title">
                                        <strong> Jhon doe</strong>
                                    </div>
                                    <div class="message-detail">Lorem ipsum dolor sit amet consectetur adipisicing
                                        elit. In totam explicabo</div>
                                </div>
                            </div>

                        </a>
                        <a class="dropdown-item" href="#">
                            <div class="message-content">
                                <div class="pic">
                                    <img src="img/user.jpg" alt="">
                                </div>
                                <div class="content">
                                    <div class="message-title">
                                        <strong> Jhon doe</strong>
                                    </div>
                                    <div class="message-detail">Lorem ipsum dolor sit amet consectetur adipisicing
                                        elit. In totam explicabo</div>
                                </div>
                            </div>
                        </a>
                        <div class="dropdown-divider"></div>
                        <a class="dropdown-item text-center" href="#">View all messages</a>

                    </div>
                </div>
                <div class="dropdown">
                    <a href="#" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                        <i class="fa fa-cog"></i>
                        <span class="badge-sonar"></span>
                    </a>
                    <div class="dropdown-menu" aria-labelledby="dropdownMenuMessage">
                        <a class="dropdown-item" href="#">My profile</a>
                        <a class="dropdown-item" href="#">Help</a>
                        <a class="dropdown-item" href="#">Setting</a>
                    </div>
                </div>
                <div>
                    <a href="#">
                        <i class="fa fa-power-off"></i>
                    </a>
                </div>
                <div class="pinned-footer">
                    <a href="#">
                        <i class="fas fa-ellipsis-h"></i>
                    </a>
                </div>
            </div> -->
        </nav>
        <!-- page-content  -->
<main class="page-content">
<div id="overlay" class="overlay"></div>
<!-- start of the current view container -->
<div id="application-current-view"> 
<div id="single_cell_lncrna_database">
    <nav class="navbar navbar-expand-lg navbar-dark bg-custom-navbar"> <!-- sticky-top -->
        <a class="navbar-brand mb-0 h1">Single Cell Long noncoding RNA Atlas</a>
        <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarNav_top" aria-controls="navbarNav_top" aria-expanded="false" aria-label="Toggle top navigation"><span class="navbar-toggler-icon"></span></button>
        <div class="collapse navbar-collapse" id="navbarNav_top">
            <ul class="navbar-nav">
                <li class="nav-item active">
                    <a class="nav-link" type="button" data-toggle="collapse" data-target="#introduction" aria-controls="introduction" aria-expanded="false" aria-label="Toggle introduction" onclick="Toggle_Introduction_Modal( )">Introduction</a>
                </li>
                <li class="nav-item active">
                    <a class="nav-link" type="button" aria-label="Toggle Side Menu" id="toggle-sidebar">Side Menu</a>
                </li>
                <li class="nav-item active">
                    <a class="nav-link" href="documentation/index.html">Documentation</a>
                </li>
                <li class="nav-item">
                    <a class="nav-link">Citation</a>
                </li>
                <li class="nav-item">
                    <a class="nav-link">Feedback</a>
                </li>
            </ul>
            <button type="button" class="btn btn-light bth-sm ml-auto mr-1 my-0" aria-label="capture current view and download it as a PNG file" data-tooltip="tooltip" data-placement="left" data-html="true" title="Capture <b>current view</b> and download it as a PNG file" onclick="DownloadCurrentView( )"><i class="material-icons icon-black" style="font-size:25px">camera_alt</i></button>
            <button type="button" class="btn btn-light bth-sm ml-1 mr-1 my-0" aria-label="show keyboard shortcuts" data-tooltip="tooltip" data-placement="bottom" data-html="true" title="Press to show <b>keyboard shortcuts</b>" onclick="$( '#modal_keyboard_shortcuts' ).modal( 'show' )"><i class="material-icons icon-black" style="font-size:25px">keyboard</i></button>
            <button type="button" class="btn btn-light bth-sm ml-1 my-0" aria-label="toggle tooltips" data-tooltip-always-active="tooltip" data-placement="bottom" data-html="true" title="Press to <b>on/off</b> help messages on mouse over (hover) of certain buttons" onclick="Toggle_Tooltips( )"><i class="material-icons icon-black" style="font-size:25px">help_outline</i></button>
        </div>
    </nav>
    <button onclick="Toggle_Sidebar( )" id="button_open_sidebar" title="Open Sidebar"><i class="fas fa-sliders-h"></i></button>
    <div id="container_single_cell_scatter_plot">
        <div id="controlBox_plotly_graph_main">
            <div class="pos-f-t">
                <nav class="navbar navbar-dark bg-custom-navbar">
                    <form class="form-inline">
                        <button class="navbar-toggler mr-auto" type="button" data-toggle="collapse" data-target="#navbarToggleExternalContent__plotly_graph_umap" aria-controls="navbarToggleExternalContent__plotly_graph_umap" aria-expanded="false" aria-label="Toggle navigation" data-tooltip="tooltip" data-placement="right" data-html="true" title="Press to change default graph settings and turn on/off tracks"><i class="material-icons icon-white" style="font-size:20px;cursor:pointer">settings</i></button>
                        <a class="mb-0" style="color:white; font-size:0.8em" id="text_number_of_records"></a>
                    </form>
                    
                    <form class="form-inline mx-auto">
                        <div class="input-group-prepend mx-2 my-0" style="height:30px"><h5 id="text_umap_annotation" class="input-group-text">Annotation</h5></div>
                        <select class="form-control my-0" id="select_umap_annotation_data" style="text-align: center; height:30px; line-height: 0px; padding-top: 0px; padding-bottom: 0px;" data-tooltip="tooltip" data-placement="bottom" data-html="true" title="Select data to be visualized on the 'Annotation' UMAP plot" onchange="PlotCellAnnotation( flag_restyle = true )">
                            <option value="cell type" selected>Cell Type</option>
                            <option value="tissue">Tissue</option>
                            <option value="nCount_RNA">Number of mRNA</option>
                            <option value="nFeature_RNA">Number of Genes</option>
                            <option value="percent_mt">MT Percentage</option>
                        </select>
<!--                         <button class="btn btn-light btn-sm mx-2 my-0" type="button" data-toggle="button" data-tooltip="tooltip" data-placement="left" data-html="true" title="Press to <b>on/off</b> the <i>legend</i> of the graph" id="button_toggle_legend_UMAP" onclick="Toggle_Legend_UMAP( )"><i class="fas fa-list"></i></button> -->
                    </form>
                    <form class="form-inline mx-auto">
                        <div class="input-group-prepend mx-2 my-0 my-0" style="height:30px"><h5 id="text_umap_expression_x" class="input-group-text">X</h5></div>
                        <select class="form-control my-0" id="select_scatter_x_axis" style="text-align: center; height:30px; line-height: 0px; padding-top: 0px; padding-bottom: 0px;" data-tooltip="tooltip" data-placement="bottom" data-html="true" title="Select X-axis of scatter plots. When selecting 'Gene', click a gene of interest in the list of gene symbol tags in the search box to set the expression of the given gene as the axis." onchange="Change_UMAP_Coordinate( null );">
                            <option value="__UMAP__" selected>UMAP-1</option>
                            <option value="Gene">Gene</option>
<!--                             <option value="Scatter">Scatter</option> -->
                        </select>
                        <div class="input-group-prepend mx-2 my-0 my-0" style="height:30px"><h5 id="text_umap_expression_y" class="input-group-text">Y</h5></div>
                        <select class="form-control my-0" id="select_scatter_y_axis" style="text-align: center; height:30px; line-height: 0px; padding-top: 0px; padding-bottom: 0px;" data-tooltip="tooltip" data-placement="bottom" data-html="true" title="Select Y-axis of scatter plots. When selecting 'Gene', click a gene of interest in the list of gene symbol tags in the search box to set the expression of the given gene as the axis." onchange="Change_UMAP_Coordinate( null );">
                            <option value="__UMAP__" selected>UMAP-2</option>
                            <option value="Gene">Gene</option>
<!--                             <option value="Scatter">Scatter</option> -->
                        </select>
<!--                         <button class="btn btn-light btn-sm mx-2 my-0" type="button" data-toggle="button" data-tooltip="tooltip" data-placement="bottom" data-html="true" title="Press to <b>on/off</b> the <i>legend</i> of the graph" id="button_umap_expr_left" onclick=""><i class="fas fa-caret-up"></i></button> -->
<!--                         <button class="btn btn-light btn-sm my-0" type="button" data-toggle="button" data-tooltip="tooltip" data-placement="bottom" data-html="true" title="Press to <b>on/off</b> the <i>legend</i> of the graph" id="button_umap_expr_right" onclick=""><i class="fas fa-caret-down"></i></button> -->
                    </form>
                </nav>
                <div class="collapse" id="navbarToggleExternalContent__plotly_graph_umap">
                    <div class="bg-custom-navbar-hidden p-4">
                        <h4 class="text-black">Main Graph Setting Panel</h4>
                        <p>
                            <h5 class="text-muted">General Setting</h5>
                            <div class="btn-toolbar" role="toolbar">
                                <div class="input-group mx-2 my-2">
                                    <div class="input-group-prepend"><span class="input-group-text">Plot height is</span></div>
                                    <input type="text" class="form-control" aria-label="height of the plot in pixel" id="input_plot_height_UMAP" title="input field for setting the height of UMAP plots in pixels" value="750">
                                    <div class="input-group-append"><span class="input-group-text">pixels</span></div>
                                </div>
                            </div>
                        </p>
                        <p>
                            <h5 class="text-muted">Violin Plot Setting</h5>
                            <div class="btn-toolbar" role="toolbar">
                                <div class="input-group mx-2 my-2">
                                    <div class="input-group-prepend"><span class="input-group-text">Plot height is</span></div>
                                    <input type="text" class="form-control" aria-label="height of the plot in pixel" id="input_plot_height_violin" title="input field for setting the height of violin plots in pixels" value="750">
                                    <div class="input-group-append"><span class="input-group-text">pixels</span></div>
                                </div>
                            </div>
                            <div class="btn-toolbar" role="toolbar">
                                <div class="input-group">
                                    <div class="form-check form-check-inline">
                                        <input id="checkbox__plotly_sclncrna_expression_violine__ignore_cells_with_zero_counts" type="checkbox" checked data-toggle="toggle" data-on="All Cells" data-off="Only Positive Cells" data-onstyle="light" data-offstyle="secondary" data-size="sm" data-style="ios" data-width="150" onchange="UpdateGene( )">
                                    </div>
                                </div>
                            </div>
                        </p>
                        <p><button type="button" class="btn btn-secondary btn-block mx-2 my-2" onclick="UpdateGene( );PlotCellAnnotation( );">Update Graphs</button></p>
                    </div>
                </div>
            </div>
        </div> 
        <div id="container_main_panel">
            <div id="container_searchbox_main">
                <input name='tags' id='searchbox_main' placeholder='Enter Gene Symbol' value='' data-blacklist='.NET,PHP'>
            </div>
            <div id="container_plotly_searchbox_main_preview_container" style="position:relative">
                <div id="container_plotly_searchbox_main_preview" class="collapse" style="position:absolute; z-index:990; left:0; top:0;">
                    <div id="plotly_searchbox_main_preview"></div>
                </div>
            </div>
            <div class="deconstructed_pancake__parent">
                <div class="deconstructed_pancake__box"><div id="plotly_sclncrna_expression_umap_annotation"></div></div>
                <div class="deconstructed_pancake__box"><div id="plotly_sclncrna_expression_umap_expression"></div></div>
            </div>
        </div>
    </div>
    <div>
        <div id="plotly_sclncrna_expression_violine"></div>
    </div>
    <nav class="navbar navbar-dark bg-custom-navbar">
        <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarToggleExternalContent_supplementary_graph_setting_panel" aria-controls="navbarToggleExternalContent_supplementary_graph_setting_panel" aria-expanded="false" aria-label="Toggle supplementary plot setting panel" data-tooltip="tooltip" data-placement="right" data-html="true" title="Press to <b>select or remove columns</b> in for <i>parallel coordinate graphs</i> for filtering cross-reactivity prediction data"><i class="material-icons icon-white" style="font-size:20px;cursor:pointer">settings</i></button>
        <form class="form-inline">
            <button class="btn btn-light btn-sm mx-1 my-0" id="btn_cell_marker_table" type="button" data-toggle="collapse" data-toggle="button" data-target="#container_panel_datatable_cell_marker" aria-expanded="false" aria-controls="container_panel_datatable_cell_marker" data-tooltip="tooltip" data-placement="bottom" data-html="true" title="Press to <b>toggle</b> <i>a data table</i> displaying <b>cell markers</b> for a given cell cluster" aria-label="">Cell Marker Table</button>
            <button class="btn btn-light btn-sm mx-1 my-0" id="btn_correl_gene_table" type="button" data-toggle="collapse" data-toggle="button" data-target="#container_panel_datatable_correl" aria-expanded="false" aria-controls="container_panel_datatable_correl" data-tooltip="tooltip" data-placement="bottom" data-html="true" title="Press to <b>toggle</b> <i>a data table</i> displaying a list of genes <b>whose expression is correlated with the current gene</b>" aria-label="">Co-expressed Gene Table</button>
<!--             <button class="btn btn-light btn-sm mx-1 my-0" type="button" data-toggle="collapse" data-toggle="button" data-target="#container_interaction_network" aria-expanded="false" aria-controls="container_interaction_network"data-tooltip="tooltip" data-placement="bottom" data-html="true" title="Press to <b>toggle</b> <i>a table</i> showing <b>predicted/experimental structural properties</b> of aligned query and SARS-CoV-2 target proteins" aria-label="">Interactions</button> -->
        </form>
    </nav>
    <div class="collapse" id="navbarToggleExternalContent_supplementary_graph_setting_panel">
        <div class="bg-custom-navbar-hidden p-4">
            <h4 class="text-black"> Supplementary Graph Setting Panel</h4>
        </div>
    </div>
    <div id="container_panel_datatable_cell_marker" class="collapse">
        <nav class="navbar navbar-dark bg-custom-navbar">
            <a class="navbar-brand mb-0 h5" id="panel_name_cellmarker">Cell Marker Table</a>
            <div class="btn-toolbar" role="toolbar">
                <div class="input-group mr-1 ml-1 my-0">
                    <button class="btn btn-light ml-1 mr-2 my-0" type="button" onclick="Download_DataTable( 'datatable_cell_marker' )" data-tooltip="tooltip" data-placement="bottom" title="Press to download current datatable as a CSV file"><i class="material-icons icon-black" style="font-size:20px;cursor:pointer">get_app</i></button> 
                    <div class="input-group-append"><span class="input-group-text">Download Data</span></div>
                </div>
            </div>
            <button type="button" class="close" aria-label="Close container_panel_datatable_cell_marker" data-toggle="collapse" data-target="#container_panel_datatable_cell_marker"><span aria-hidden="true">&times;</span></button>
        </nav>
        <div id="container_datatable_cell_marker"><table id="datatable_cell_marker" class="display" width="100%"></table></div>
    </div>
    <div id="container_panel_datatable_correl" class="collapse">
        <nav class="navbar navbar-dark bg-custom-navbar">
            <a class="navbar-brand mb-0 h5" id="panel_name_correl_gene">Co-expression Table</a>
            <div class="btn-toolbar" role="toolbar">
                <div class="input-group mr-1 ml-1 my-0">
                    <button class="btn btn-light ml-1 mr-2 my-0" type="button" onclick="Download_DataTable( 'datatable_correl' )" data-tooltip="tooltip" data-placement="bottom" title="Press to download current datatable as a CSV file"><i class="material-icons icon-black" style="font-size:20px;cursor:pointer">get_app</i></button> 
                    <div class="input-group-append"><span class="input-group-text">Download Data</span></div>
                </div>
            </div>
            <button type="button" class="close" aria-label="Close container_panel_datatable_correl" data-toggle="collapse" data-target="#container_panel_datatable_correl"><span aria-hidden="true">&times;</span></button>
        </nav>
        <div id="container_datatable_correl"><table id="datatable_correl" class="display" width="100%"></table></div>
    </div>
<!--     <div id="container_interaction_network" class="collapse">
        <nav class="navbar navbar-dark bg-custom-navbar">
            <a class="navbar-brand mb-0 h5">Data Tables</a>
            <div class="btn-toolbar" role="toolbar">
                <div class="input-group mr-2 ml-1 my-0">
                    <select class="custom-select w-auto" id="select_datatable" title="select datatable">
                        <option value="B-cell cross-reactivity prediction">select datatable</option>
                        <option value="B-cell cross-reactivity prediction" selected>B-cell cross-reactivity prediction</option>
                        <option value="T-cell cross-reactivity prediction">T-cell cross-reactivity prediction</option>
                        <option value="query proteins">query proteins</option>
                        <option value="target proteins">target proteins</option>
                        <option value="MHC alleles">MHC alleles</option>
                    </select>
                    <div class="input-group-append"><span class="input-group-text">datatable</span></div>
                </div>
                <button type="button" class="btn btn-light mx-0 my-0" id="button_display_datatable" aria-pressed="false" onclick="Display_DataTable( )" autocomplete="off">Show</button>
                <button class="btn btn-light ml-1 mr-2 my-0" type="button" onclick="Download_DataTable( )" data-tooltip="tooltip" data-placement="bottom" title="Press to download current datatable as a CSV file"><i class="material-icons icon-black" style="font-size:20px;cursor:pointer">get_app</i></button> 
            </div>
            <button type="button" class="close" aria-label="Close container_interaction_network" data-toggle="collapse" data-target="#container_interaction_network"><span aria-hidden="true">&times;</span></button>
        </nav>
        <div id="interaction_network" style="height:100px;">
            <div>
                <h1>lncRNA Interaction Graphs</h1>
                <label>
                  Filter nodes
                  <select id="nodeFilterSelect">
                    <option value>All characters</option>
                    <option value="kid">kids</option>
                    <option value="adult">adults</option>
                    <option value="male">male</option>
                    <option value="female">female</option>
                  </select>
                </label>
                <br />
                <br />
                <label>
                  Filter edges
                  <div>
                    <label>
                      <input type="checkbox" name="edgesFilter" value="parent" checked />
                      Is <span style="color: green">parent</span> of
                    </label>
                  </div>
                  <div>
                    <label>
                      <input type="checkbox" name="edgesFilter" value="teacher" checked />
                      Is <span style="color: blue">teacher</span> of
                    </label>
                  </div>
                  <div>
                    <label>
                      <input type="checkbox" name="edgesFilter" value="friend" checked />
                      Is <span style="color: red">friend</span> of
                    </label>
                  </div>
                </label>
                <br />
                <div id="mynetwork"></div>
            </div>
        </div>
    </div> -->

    

</div>
</div> 
<!-- end of current view container -->
</main>
<!-- page-content" -->
</div>


<div class="modal fade" id="modal_keyboard_shortcuts" tabindex="-1" role="dialog" aria-labelledby="modal_keyboard_shortcuts" aria-hidden="true">
    <div class="modal-dialog modal-lg" role="document">
        <div class="modal-content">
            <div class="modal-header">
                <h5 class="modal-title" id="label_modal_generate_peptides">Keyboard Shortcuts</h5>
                <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
            </div>
            <div class="modal-body">
                <h5 class="text-muted mb-2" style="font-size:20px">General</h5>
                <p>Press <kbd>Ctrl</kbd> or <kbd>Cmd</kbd> + <kbd>s</kbd><b>(screenshot)</b> to capture <b>the current view</b> and download it as a PNG file</p>
                <p>Press <kbd>Ctrl</kbd> or <kbd>Cmd</kbd> + <kbd>i</kbd><b>(introduction)</b> to toggle the introduction and instructions</p>
                <p>Press <kbd>Ctrl</kbd> or <kbd>Cmd</kbd> + <kbd>h</kbd><b>(help messages)</b> to toggle tooltip messages that are shown when mouse pointers are on buttons.</p>
                <p>Press <kbd>Ctrl</kbd> or <kbd>Cmd</kbd> + <kbd>u</kbd><b>(UP)</b> to move to the top of the page.</p>
                <p>Press <kbd>Ctrl</kbd> or <kbd>Cmd</kbd> + <kbd>d</kbd><b>(Down)</b> to move to the bottom of the page.</p>
                <br>
            </div>
            <div class="modal-footer">
                <button type="button" class="btn btn-secondary" data-dismiss="modal">Okay</button>
            </div>
        </div>
    </div>
</div>
<!-- fullscreen introduction modal -->
<div id="modal_introduction" class="modal fade show" tabindex="-1" role="dialog" aria-labelledby="modal_introduction" aria-hidden="true" style="padding-right: 0px!important;"> <!-- disable fading when the web page has been loaded -->
  <div class="modal-dialog" style="width: 100%!important;height: 100%!important;margin: 0!important;padding: 0!important;max-width:none !important;">
    <div class="modal-content" style="height: auto !important; min-height: 100% !important; border-radius: 0 !important;min-width: 100% !important;">
        <nav class="navbar navbar-expand-lg navbar-dark bg-custom-navbar"> <!-- sticky-top -->
            <a class="navbar-brand mb-0 h1">Single Cell Long noncoding RNA Atlas</a>
            <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarNav_top" aria-controls="navbarNav_top" aria-expanded="false" aria-label="Toggle top navigation"><span class="navbar-toggler-icon"></span></button>
            <div class="collapse navbar-collapse" id="navbarNav_top">
                <ul class="navbar-nav">
                    <li class="nav-item">
                        <a class="nav-link">Citation</a>
                    </li>
                    <li class="nav-item">
                        <a class="nav-link">Feedback</a>
                    </li>
                </ul>
            </div>
        </nav>
            <div class="modal-header">
                <h1 class="modal-title">Introduction</h1>
                <button class="btn btn-link btn-lg mt-2 mr-auto" onclick="Toggle_Introduction_Modal();">Skip Introduction</button>
                <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
            </div>
            <div class="modal-body">
                <h2></h2>
                
                <h2></h2>
                <img src="img/abstract.jpg" alt="introduction images" style="min-width:50%;max-width:90%;display: block;margin-left: auto;margin-right: auto;">
                <br>
                <div class="card mx-auto my-2" style="max-width: 60rem;">
                    <div class="card-body">
                        <h2 class="card-title">Abstract</h2>
                        <p class="card-text">Accumulated evidence demonstrates that long non-coding RNAs (lncRNAs) regulate cell differentiation and homeostasis, influencing kidney aging and disease. Despite their versatility, the function of lncRNA remain poorly understood due to the lack of a reference map of lncRNA transcriptome in various cell types. In this study, we employed a targeted single-cell RNA sequencing (scRNA-seq) method to enrich and characterize lncRNAs in individual cells. We applied this method to various mouse tissues, including normal and aged kidneys, and generated a comprehensive single-cell lncRNA atlas (https://gist-fgl.github.io/sc-lncrna-atlas/). Through tissue-specific clustering analysis, we identified cell type-specific lncRNAs that showed a high correlation with known cell-type marker genes. Furthermore, we constructed gene regulatory networks (GRNs) to explore the functional roles of differentially expressed lncRNAs in each cell type. In the kidney, we observed dynamic expression changes of lncRNAs during aging, with specific changes in glomerular cells. These cell type- and age-specific expression patterns of lncRNAs provide insights into their potential roles in regulating cellular processes, such as immune response and energy metabolism, during kidney aging. Our study highlights the importance of lncRNAs in orchestrating gene expression networks in a cell type-specific manner and provides a valuable resource for further investigations into the functional significance of lncRNAs in diverse biological contexts.</p>
                    </div>
<!--                     <div class="mr-auto ml-auto"><a style="font-size: 1.35em; font-weight: bold" target="_blank" rel="noopener noreferrer" class="btn btn-link btn-sm mx-1 my-0" role="button" data-toggle="button" data-tooltip="tooltip" data-placement="bottom" data-html="true" title="Click to view our BioRXiv preprint">BioRXiv preprint<i class="material-icons icon-black" style="font-size: 0.9em">launch</i></a></div> -->
                </div>

<!--                 <h1 class="modal-title">Usages</h1> -->

                <h5 class="text-muted" style="text-align:center;">Data analysis has been conducted by Gyeongdae Kim and the web application has been developed by Hyunsu An at Gwangju Institute of Science and Technology under the supervision of Professor JiHwan Park.</h5>
            </div>
            <div class="modal-footer">
                <button class="btn btn-secondary btn-lg btn-block" data-dismiss="modal">Start!</button>
            </div>
        </div>
    </div>
</div>
<div id="loading_screen"></div>

<script src="js/github__pro_sidebar-jquery.mCustomScrollbar.concat.min.js"></script>
<script src="js/github__pro_sidebar-main.js"></script>
    
<script>
// packages
var DragSort=function(t){
    var e,i=0,s={},r={},a=(e=window.MutationObserver||window.WebKitMutationObserver,function(t,i){t&&1===t.nodeType&&(e?new e((function(t,e){i(t)})).observe(t,{childList:!0,subtree:!1}):window.addEventListener&&t.addEventListener("DOMNodeInserted",i,!1))});function n(t,e){if(!t)return this;e=e||{},this.parentElm=t,this.uid=e.uid,this.settings={selector:"*",callbacks:{}},Object.assign(this.settings,e),this.setup(),a(this.parentElm,this.setup.bind(this)),this.bindEvents()}return n.prototype={namespace:"dragsort",setup(){[...this.parentElm.childNodes].forEach(t=>{if(1!=t.nodeType)return t.parentNode.removeChild(t);t.matches(this.settings.selector)&&(t.draggable=!0)}),this.gap=this.getItemsGap(this.parentElm.firstElementChild)},throttle(t,e){var i=!1,s=this;return function(r){i||(t.call(s,r),i=!0,setTimeout(()=>i=!1,e))}},getDraggableElm(t){var e=t.closest('[draggable="true"]');return this.uid==s.uid?e:null},dragstart(t,e){s=this;var i,r=this.getDraggableElm(e);r?(this.source=this.getInitialState(),this.target=this.getInitialState(),i=r.getBoundingClientRect(),this.source.elm=r,this.source.idx=this.getNodeIndex(r),this.source.size.width=i.width,this.source.size.height=i.height,t.dataTransfer.effectAllowed="move",setTimeout(this.afterDragStart.bind(this))):s={}},afterDragStart(){var t="vertical"==this.settings.mode?"height":"width";this.parentElm.classList.add(this.namespace+"--dragStart"),this.source.elm.style[t]=this.source.size[t]+"px",this.source.elm.classList.add(this.namespace+"--dragElem")},dragover(t){t.preventDefault(),t.stopPropagation();var e=t.target;if((e=this.getDraggableElm(e))&&this.target){var i=this.target.elm,s=this.target.hoverDirection;t.dataTransfer.dropEffect="move",this.target.hoverDirection=this.getTargetDirection(t),i==e&&s==this.target.hoverDirection||this.directionAwareDragEnter(t,e)}},dragenter(t,e){(e=this.getDraggableElm(e))&&this.target&&this.isValidElm(e)&&this.source.elm!=e&&this.source.elm&&(this.target.bounding=e.getBoundingClientRect())},directionAwareDragEnter(t,e){var i;t.preventDefault(),t.stopPropagation(),t.dataTransfer.dropEffect="none",this.isValidElm(e)&&this.source.elm!=e&&this.source.elm&&(t.dataTransfer.dropEffect="move",this.cleanupLastTarget(),this.target.elm=e,this.target.idx=this.getNodeIndex(e),e.classList.add("over"),i=Math.abs(this.target.idx-this.source.idx),this.source.elm.classList.toggle(this.namespace+"--hide",i>0),"vertical"==this.settings.mode?this.target.elm.style[this.target.hoverDirection?"marginBottom":"marginTop"]=this.source.size.height+this.gap+"px":this.target.elm.style[this.target.hoverDirection?"marginRight":"marginLeft"]=this.source.size.width+this.gap+"px")},dragend(t){if(clearTimeout(this.dragoverTimeout),this.dragoverTimeout=null,this.parentElm.classList.remove(this.namespace+"--dragStart"),!this.isValidElm(this.target.elm))return this.cleanup();var e=this.target.hoverDirection?this.target.elm.nextElementSibling:this.target.elm;return this.source.elm!=this.target.elm&&this.target.elm&&(this.target.elm.classList.add(this.namespace+"--noAnim"),this.cleanup(),this.parentElm.insertBefore(this.source.elm,e)),this.source.elm&&this.source.elm.classList.remove(this.namespace+"--dragElem",this.namespace+"--hide"),this.settings.callbacks.dragEnd(this.source.elm),this},isTargetLastChild(){return this.parentElm.lastElementChild==this.target.elm},getTargetDirection(t){if(this.target.bounding)return"vertical"==this.settings.mode?t.pageY>this.target.bounding.top+this.target.bounding.height/2?1:0:t.pageX>this.target.bounding.left+this.target.bounding.width/2?1:0},getNodeIndex(t){for(var e=0;t=t.previousSibling;)3==t.nodeType&&/^\s*$/.test(t.data)||e++;return e},isValidElm(t){return t&&t.nodeType&&t.parentNode==this.parentElm},cleanup(){s={},[...this.parentElm.children].forEach(t=>{t.removeAttribute("style"),setTimeout(()=>{t.classList.remove(this.namespace+"--over",this.namespace+"--noAnim",this.namespace+"--dragElem")},50)})},cleanupLastTarget(){this.target.elm&&(this.target.elm.classList.remove(this.namespace+"--hide",this.namespace+"--over"),this.target.elm.removeAttribute("style"))},getInitialState:()=>({elm:null,size:{}}),getItemsGap(t){var e=getComputedStyle(t);return"vertical"==this.settings.mode?parseInt(e.marginTop)+parseInt(e.marginBottom):parseInt(e.marginLeft)+parseInt(e.marginRight)},bindEvents(t){for(var e in this.listeners=this.listeners||{dragstart:t=>this.dragstart(t,t.target),dragenter:t=>this.dragenter(t,t.target),dragend:t=>this.dragend(t,t.target),dragover:this.throttle(this.dragover,350)},this.listeners)this.parentElm[t?"removeEventListener":"addEventListener"](e,this.listeners[e])},destroy(){this.cleanup(),this.bindEvents(!0),delete r[this.uid]}},function(t,e){return r[++i]=t._DragSort?r[t._DragSort]:new n(t,{...e,uid:i}),t._DragSort=i,r[i]}}();

// Initialization
    var name_application = "Single_Cell_LncRNA_Database"; // define name of the application
    $(document).ready( function( ) {
        $('[data-toggle="tooltip"]').tooltip( ); // enable tooltips
        $('[data-tooltip="tooltip"]').tooltip( );
        $('[data-tooltip-always-active="tooltip"]').tooltip( );
        $( '#modal_introduction' ).modal( 'show' ); // show introduction modal
        Toggle_Tooltips( ); // set tooltips status
        Initialize( );
    });
    // settings
    var object_data = { }; // an object containing data
    object_data.file = { }; 
    object_data.file.base_url = "current/rawdata/base64/";
    object_data.file.dict_name_file_to_parsed_data = { };
    object_data.main = { };
    object_data.main.gene = { };
    object_data.main.combined = { 'Combined.Aging_Kidney' : { 'Glom (Old)' : 1, 'Glom (Young)' : 1, 'Tubule (Old)' : 1, 'Tubule (Young)' : 1 }, 'Combined.mTissues6' : { 'Heart' : 1, 'Intestine' : 1, 'Kidney' : 1, 'Lung' : 1, 'Liver' : 1, 'Thymus' : 1 } }; // define combined dataset
    object_data.main.filters_for_subsets = { }; // an empty object that will contains filters for subset
    object_data.setting = { };
    object_data.setting.app = { };
    object_data.setting.app.name_tissue_default = 'Combined.Aging_Kidney'; // name_tissue of the initial dataset that will be loaded by default
    object_data.setting.plot = { };
    object_data.flags = { };
    object_data.main.gene.expr = { data : { }, data_size : 250 }; // cache of expression data // size of the cache
    object_data.main.cell_marker = { data : { }, data_size : 5 };
    object_data.main.gene.correl = { };
    object_data.main.gene.correl.dict_name_tissue_to_prefix = { 
        'Heart' : 'expr.correlation.tissue.Heart', 
        'Intestine' : 'expr.correlation.tissue.Intestine', 
        'Kidney' : 'expr.correlation.tissue.Kidney', 
        'Lung' : 'expr.correlation.tissue.Lung', 
        'Liver' : 'expr.correlation.tissue.Liver', 
        'Thymus' : 'expr.correlation.tissue.Thymus', 
        'Glom (Old)' : 'expr.correlation.tissue.Kidney_Glom_Old', 
        'Tubule (Old)' : 'expr.correlation.tissue.Kidney_Tubule_Old', 
        'Glom (Young)' : 'expr.correlation.tissue.Kidney_Glom_Young', 
        'Tubule (Young)' : 'expr.correlation.tissue.Kidney_Tubule_Young', 
        'Combined.mTissues6' : 'expr.correlation.combined.mTissues6', 
        'Combined.Aging_Kidney' : 'expr.correlation.combined.Aging_and_Young_Kidney_Glom_Enrich', 
    }
    object_data.main.gene.correl.dict_name_tissue_to_cache = { };
    object_data.main.umap = { tissue : 'not_drawn' }; // set default values
    object_data.main.scatter = { x_label : 'UMAP-1', y_label : 'UMAP-2', x_axis_option : '__UMAP__', y_axis_option : '__UMAP__', x_axis_col_name : 'UMAP_1', y_axis_col_name : 'UMAP_2', flag_waiting_selection_of_gene : false, gene_symbol_selected : '', flag_axis_using_gene_expr : false }; // initialize axis of the scatter plot
    object_data.setting.expression = { };
    object_data.setting.expression.n_digits_below_floating_point = 3; // number of digits below floading point of the original expression values (converted to integer by multiplying 1000 during delivery of the data)
    object_data.main.datatable = { 'datatable_cell_marker' : { df : null, name : 'cell_marker' }, 'datatable_correl' : { df : null, name : 'correl_gene' } };
    
    object_data.setting.plot.annotation = { };
    object_data.setting.plot.annotation.dict_tissue_to_dict_cell_type_to_color = {
        'Combined.Aging_Kidney': {'podocyte': '#A63603', 'mesangial': '#FDBE85', 'JGC': '#E6550D', 'PEC': '#FD8D3C', 'PT1': '#A1D99B', 'PT2': '#238B45', 'ALOH': '#4292C6', 'CD PC': '#9ECAE1', 'GEC': '#FEB24C', 'endo_podo': '#08306B', 'endo_ALOH': '#4292C6', 'endo_immune': '#2171B5', 'lymphocytes': '#54278F', 'myeloid': '#9E9AC8', 'RBC': '#FCCDE5'}, 
        'Glom (Old)': {'podocyte': '#A63603', 'mesangial': '#FDBE85', 'JGC': '#E6550D', 'PEC': '#FD8D3C', 'PT1': '#A1D99B', 'PT2': '#238B45', 'ALOH': '#4292C6', 'CD PC': '#9ECAE1', 'GEC': '#FEB24C', 'endo_podo': '#08306B', 'endo_ALOH': '#4292C6', 'endo_immune': '#2171B5', 'lymphocytes': '#54278F', 'myeloid': '#9E9AC8', 'RBC': '#FCCDE5'}, 
        'Glom (Young)': {'podocyte': '#A63603', 'mesangial': '#FDBE85', 'JGC': '#E6550D', 'PEC': '#FD8D3C', 'PT1': '#A1D99B', 'PT2': '#238B45', 'ALOH': '#4292C6', 'CD PC': '#9ECAE1', 'GEC': '#FEB24C', 'endo_podo': '#08306B', 'endo_ALOH': '#4292C6', 'endo_immune': '#2171B5', 'lymphocytes': '#54278F', 'myeloid': '#9E9AC8', 'RBC': '#FCCDE5'}, 
        'Tubule (Old)': {'podocyte': '#A63603', 'mesangial': '#FDBE85', 'JGC': '#E6550D', 'PEC': '#FD8D3C', 'PT1': '#A1D99B', 'PT2': '#238B45', 'ALOH': '#4292C6', 'CD PC': '#9ECAE1', 'GEC': '#FEB24C', 'endo_podo': '#08306B', 'endo_ALOH': '#4292C6', 'endo_immune': '#2171B5', 'lymphocytes': '#54278F', 'myeloid': '#9E9AC8', 'RBC': '#FCCDE5'}, 
        'Tubule (Young)': {'podocyte': '#A63603', 'mesangial': '#FDBE85', 'JGC': '#E6550D', 'PEC': '#FD8D3C', 'PT1': '#A1D99B', 'PT2': '#238B45', 'ALOH': '#4292C6', 'CD PC': '#9ECAE1', 'GEC': '#FEB24C', 'endo_podo': '#08306B', 'endo_ALOH': '#4292C6', 'endo_immune': '#2171B5', 'lymphocytes': '#54278F', 'myeloid': '#9E9AC8', 'RBC': '#FCCDE5'}, 
        'Combined.mTissues6': {'PT': 'yellowgreen', 'hepatocyte1': 'darkorchid', 'hepatocyte2': '#FF0000', 'kidney_other_tubule': '#FDB462', 'Cardiomyocyte': '#27408B', 'Proli_ISC': '#CD5555', 'Ciliated': '#EEB422', 'AT1': '#CD2626', 'AT2': '#90EE90', 'Fibroblast2': '#EE4000', 'Fibroblast1': '#8DD3C7', 'Fibro3': '#8B1C62', 'podocyte': '#66C2A5', 'mesangial': '#FC8D62', 'JGC': '#8DA0CB', 'Pericyte': '#E78AC3', 'Myofibro': '#A6D854', 'GEC': '#FFD92F', 'Endo1': '#E5C494', 'Endo2': '#B3B3B3', 'LE': '#8DD3C7', 'DN_DP': '#BC80BD', 'T cell': '#B3CDE3', 'NK': '#FB8072', 'B cell': '#80B1D3', 'B cell2': '#FDB462', 'plasma1': '#B3DE69', 'Kupffer': '#FCCDE5', 'Myeloid1': '#D9D9D9', 'AM': '#BC80BD', 'myeloid': '#CCEBC5', 'RBC': '#FFED6F'}, 
        'Kidney': {'podocyte': '#8DD3C7', 'mesangial': '#B3B3B3', 'JGC': '#E5C494', 'PEC': '#FB8072', 'PT1': '#80B1D3', 'PT2': '#FDB462', 'PST': '#B3DE69', 'ALOH': '#FCCDE5', 'DCT': '#FFD92F', 'PC': '#BC80BD', 'IC': '#CCEBC5', 'GEC': '#FFED6F', 'aterioles': '#66C2A5', 'B cell': '#FC8D62', 'CD8': '#8DA0CB', 'Macro': '#E78AC3', 'RBC': '#A6D854'}, 
        'Heart': {'Cardiomyocyte': '#B15928', 'Endo1': '#8DD3C7', 'Endo2': '#B3B3B3', 'Endo3': '#E5C494', 'LE': '#FB8072', 'Fibro1': '#80B1D3', 'Fibro2': '#FDB462', 'Fibro3': '#B3DE69', 'Sox10_Fibro': '#FCCDE5', 'Myofibro': '#FFD92F', 'SMC': '#BC80BD', 'Pericyte': '#CCEBC5', 'B cell': '#FFED6F', 'T cell': '#66C2A5', 'RM': '#FC8D62', 'RM2': '#8DA0CB', 'DC': '#E78AC3', 'Granulocyte': '#A6D854', 'Platelet': '#A6CEE3', 'RBC': '#1F78B4'}, 
        'Intestine': {'Proli_ISC': '#6A3D9A', 'Proli': '#B15928', 'VTT': '#8DD3C7', 'VTT2': '#B3B3B3', 'Endo1': '#E5C494', 'Endo2': '#FB8072', 'Endo3': '#80B1D3', 'Endo4': '#FDB462', 'LV': '#B3DE69', 'Pdgfrb+Fibro': '#FCCDE5', 'Fibro': '#FFD92F', 'Myofibro1': '#BC80BD', 'Myofibro2': '#CCEBC5', 'proli_B cell': '#FFED6F', 'B cell': '#66C2A5', 'plasma': '#FC8D62', 'plasma2': '#8DA0CB', 'Naive_Cd4': '#E78AC3', 'Cd4 T': '#A6D854', 'Naive_Cd8': '#A6CEE3', 'Cd8 T': '#1F78B4', 'NK': '#B2DF8A', 'Myeloid': '#33A02C', 'Macro': '#FB9A99', 'DC': '#E31A1C'}, 
        'Liver': {'hepatocyte1': '#6A3D9A', 'hepatocyte2': '#B15928', 'Stellate cell': '#8DD3C7', 'Endo1': '#B3B3B3', 'Endo2': '#E5C494', 'Endo3': '#FB8072', 'Endo4': '#80B1D3', 'Endo5': '#FDB462', 'B cell': '#B3DE69', 'T cell': '#FCCDE5', 'T cell2': '#FFD92F', 'T cell3': '#BC80BD', 'NK/T': '#CCEBC5', 'NK': '#FFED6F', 'Cycling': '#66C2A5', 'Kupffer': '#FC8D62', 'Mono': '#8DA0CB', 'Mono2': '#E78AC3', 'pat.Mono': '#A6D854', 'Neutro': '#A6CEE3', 'Granulocyte': '#1F78B4', 'DC': '#B2DF8A', 'RBC': '#33A02C'}, 
        'Lung': {'AT1': '#8DD3C7', 'AT2': '#B3B3B3', 'Ciliated': '#E5C494', 'Gcap': '#FB8072', 'Fibroblast1': '#80B1D3', 'Fibroblast2': '#FDB462', 'Fibroblast3': '#B3DE69', 'Myofibro': '#FCCDE5', 'Mesothelial': '#FFD92F', 'SMC2': '#BC80BD', 'Immune stromal cell': '#CCEBC5', 'Pericyte': '#FFED6F', 'Naive T': '#66C2A5', 'CD8': '#FC8D62', 'NK': '#8DA0CB', 'B cell': '#E78AC3', 'B cell2': '#A6D854', 'AM': '#A6CEE3', 'Basophils': '#1F78B4', 'DC': '#B2DF8A', 'Monocyte2': '#33A02C', 'Monocyte3': '#FB9A99', 'Neutrophils': '#E31A1C', 'proliferating': '#FDBF6F', 'Myeloid': '#FF7F00', 'RBC': '#CAB2D6'}, 
        'Thymus': {'DN': '#6A3D9A', 'ISP': '#B15928', 'DP(P)': '#8DD3C7', 'DP2(P)': '#FB8072', 'DP': '#80B1D3', 'alpha beta T': '#FDB462', 'CD8 alpha alpha': '#B3DE69', 'Cd8 SP': '#FCCDE5', 'Cd4 SP': '#FFD92F', 'Cd4 SP(M)': '#BC80BD', 'NK': '#CCEBC5', 'DC': '#FC8D62'}
    }
    object_data.setting.plot.annotation.dict_tissue_to_color = {
        "Heart" : '#F8766D', 
        "Intestine" : '#B79F00', 
        "Kidney" : "#00BA38", 
        "Lung" : "#619CFF", 
        "Liver" : "#00BFC4", 
        "Thymus" : "#F564E3", 
        "Glom (Young)" : "#7111F0",
        "Glom (Old)" : "#E590E8",
        "Tubule (Young)" : "#E6E627",
        "Tubule (Old)" : "#E8D190",
    }; // define color for each tissue
    object_data.setting.plot.preview_gene = { };
    object_data.setting.plot.preview_gene.dict_category_to_color = {
        "Heart (n=5979)" : '#F8766D',
        "Intestine (n=2330)" : '#B79F00',
        "Kidney (n=2455)" : "#00BA38",
        "Lung (n=4375)" : "#619CFF",
        "Liver (n=4191)" : "#00BFC4",
        "Thymus (n=2914)" : "#F564E3",
        "Glom (Young) (n=7658)" : "#7111F0",
        "Glom (Old) (n=6505)" : "#E590E8",
        "Tubule (Young) (n=3326)" : "#E6E627",
        "Tubule (Old) (n=2772)" : "#E8D190",
    }; // define color for each tissue
    object_data.setting.plot.preview_gene.flag_preview_on_hover = true; //setting
    
    // for faster async loading, the data files will be loaded together
    var l_remote_file = [ "current/rawdata/base64/cells.all.tsv.gz.base64.txt",
                          "current/rawdata/base64/all.box_precomputed_by_tissue.multiplied_by_1000.tsv.gz.base64.txt",
                          "current/rawdata/base64/expr.multiplied_by_1000.index.tsv.gz.base64.txt",
                          "current/rawdata/base64/cell_marker.index.tsv.gz.base64.txt",
                          `current/rawdata/base64/${object_data.main.gene.correl.dict_name_tissue_to_prefix[ object_data.setting.app.name_tissue_default ]}.index.tsv.gz.base64.txt`
                        ]; 
    
    async function Initialize( ) { // initialize the web application
        await LoadRemoteData( l_remote_file ); // load remote core data 
        console.log( "[Initialize] core data loaded", Object.keys( object_data.file.dict_name_file_to_parsed_data ) );
        await ParseData( );
        console.log( "[Initialize] processing of parsed data completed:", Object.keys( object_data.file.dict_name_file_to_parsed_data ) );
        Initialize_SearchBox_Main( );
        console.log( "[Initialize] the main Gene_Symbol search box is now ready to receive queries", Object.keys( object_data.file.dict_name_file_to_parsed_data ) );
        object_data.main.gene.gene_symbol = 'Col15a1'; // plot one gene as an example
        Change_UMAP_Coordinate( object_data.setting.app.name_tissue_default ); // load the default name_tissue
        tagify_searchbox_main.addTags( 'Col15a1') // add another tags
        PreviewGene( object_data.main.gene.gene_symbol );
        Loading_Screen( false );
    }    
        
// Bootstrap component control
    var flag_tooltip_active_Toggle_Tooltips = true;
    function Toggle_Tooltips(  ) {
        console.log( "[Toggle_Tooltips] Tooltips toggled to ", flag_tooltip_active_Toggle_Tooltips )
        flag_tooltip_active_Toggle_Tooltips = ! flag_tooltip_active_Toggle_Tooltips; // toggle the flag
        if ( flag_tooltip_active_Toggle_Tooltips ) { $('[data-tooltip="tooltip"]').tooltip( 'enable' ); }
        else { $('[data-tooltip="tooltip"]').tooltip( 'disable' ); }
    }
// load packages
    // use dataframe-js
    var DataFrame = dfjs.DataFrame;
    
// key interactions
    object_data_keydown = { dict_status: { } };
    function Record_KeyDown( event ) { // Record pressed keyborad keys, and trigger appropriate function when key is pressed.
        object_data_keydown.dict_status[ event.key ] = event.type; // record event.type for each key
    }
    window.addEventListener( "keyup", Record_KeyDown );
    window.addEventListener( "keypress", Record_KeyDown );
    window.addEventListener( "keydown", Record_KeyDown );
    hotkeys( 'ctrl+d,ctrl+i,ctrl+u,ctrl+h,command+h,command+d,command+i,command+u', function ( event, handler ) {
        event.preventDefault( );
        switch ( handler.key ) {
            case 'ctrl+d': // scroll to the bottom of the page
            case 'command+d': 
                Scroll_to_the_Bottom( ); 
                break;
            case 'ctrl+h': 
            case 'command+h': 
                Toggle_Tooltips( ); 
                break;
            case 'ctrl+i': 
            case 'command+i': 
                Toggle_Introduction_Modal( ); 
                break;
            case 'ctrl+u': 
            case 'command+u': 
                Scroll_to_the_Top( ); 
                break;
            default: alert(event);
        }
    }); 
// scroll control
    function Scroll_to_the_Top( ) {
      document.body.scrollTop = 0; // For Safari
      document.documentElement.scrollTop = 0; // For Chrome, Firefox, IE and Opera
    }
    function Scroll_to_the_Bottom( ) {
      $( document ).scrollTop( $( document ).height( ) ); // scroll to the bottom
    }
// loading screen and modal control
    function Loading_Screen( bool_add_loading_screen ) { // add or remove loading screen
        if ( ( ! bool_add_loading_screen ) && document.getElementById( "svg_during_loading" ) != null ) {
            document.getElementById( "svg_during_loading" ).remove( ); // remove loading screen once plot is drawn
        } else if ( bool_add_loading_screen && document.getElementById( "svg_during_loading" ) == null ) {
            $("#loading_screen").append( [
                '<div class="loader loader--style1 _plotly-loading-callback" title="data is being loaded..." id="svg_during_loading">',
                    '<svg version="1.1" id="loader-1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px" width="40px" height="40px" viewBox="0 0 40 40" enable-background="new 0 0 40 40" xml:space="preserve" rect="fill: #FF6700;" path>',
                        '<path opacity="0.2" fill="#000" d="M20.201,5.169c-8.254,0-14.946,6.692-14.946,14.946c0,8.255,6.692,14.946,14.946,14.946',
                        's14.946-6.691,14.946-14.946C35.146,11.861,28.455,5.169,20.201,5.169z M20.201,31.749c-6.425,0-11.634-5.208-11.634-11.634',
                        'c0-6.425,5.209-11.634,11.634-11.634c6.425,0,11.633,5.209,11.633,11.634C31.834,26.541,26.626,31.749,20.201,31.749z"/>',
                        '<path fill="#000" d="M26.013,10.047l1.654-2.866c-2.198-1.272-4.743-2.012-7.466-2.012h0v3.312h0',
                        'C22.32,8.481,24.301,9.057,26.013,10.047z">',
                        '<animateTransform attributeType="xml"',
                          'attributeName="transform"',
                          'type="rotate"',
                          'from="0 20 20"',
                          'to="360 20 20"',
                          'dur="0.5s"',
                          'repeatCount="indefinite"/>',
                        '</path>',
                    '</svg>',
                '</div>' ].join( "" ) ); // display an SVG file during loading 
        }
    }
    function Toggle_Introduction_Modal( ) { 
        $( '#modal_introduction' ).modal( 'toggle' );
        document.getElementById( "modal_introduction" ).style[ "padding-right" ] = "0px";
    } // Toggle Modal for an Introduction 
// custom utility functions
    // getElementById
    function $id( id ) {
        return document.getElementById( id );
    }
    function isString ( value ) { // Returns if a value is a string
        return typeof value === 'string' || value instanceof String;
    }
    function String_Check_Substring( string, l_substring ) {
        let flag_contains_substring = false;
        string = string.toLowerCase( );
        for ( let i = 0; i < l_substring.length; i ++ ) {
            let substring = l_substring[ i ].toLowerCase( );
            if ( string.includes( substring ) ) {
                flag_contains_substring = true;
                break;
            }
        }
        return flag_contains_substring;
    }
    // functions for numpy-like operations
    function ArraySum( l ) { // return the sum of values of an array
        return l.reduce( (a, b) => a + b, 0 );
    }
    function ArrayIndexing( l, l_index ) {
        let l_output = [ ];
        for ( let i = 0; i < l_index.length; i++ ) { l_output.push( l[ l_index[ i ] ] ); }
        return l_output
    }
    function ArrayBooleanIndexing( l, l_bool ) {
        let l_output = [ ];
        for ( let i = 0; i < l_bool.length; i++ ) { if ( l_bool[ i ] ) { l_output.push( l[ i ] ); } }
        return l_output
    }
    function ArrayBoolAND( l_l ) {
        let l_output = [ ];
        for ( let j = 0; j < l_l[ 0 ].length; j++ ) {
            let output = true;
            for ( let i = 0; i < l_l.length; i++ ) { output = output && l_l[ i ][ j ]; }
            l_output.push( output );
        }
        return l_output;
    }
    function ArrayBoolOR( l_l ) {
        let l_output = [ ];
        for ( let j = 0; j < l_l[ 0 ].length; j++ ) {
            let output = false;
            for ( let i = 0; i < l_l.length; i++ ) { output = output || l_l[ i ][ j ]; }
            l_output.push( output );
        }
        return l_output;
    }
    function ArrayBoolNOT( l ) {
        let l_output = [ ];
        for ( let i = 0; i < l.length; i++ ) { l_output.push( l[ i ] ? false : true ); }
        return l_output;
    }
    function ArrayGreaterThan( l, thres ) {
        let l_output = [ ];
        for ( let i = 0; i < l.length; i++ ) { l_output.push( l[ i ] > thres ); }
        return l_output;
    }
    function ArrayLessThan( l, thres ) {
        let l_output = [ ];
        for ( let i = 0; i < l.length; i++ ) { l_output.push( l[ i ] < thres ); }
        return l_output;
    }
    function ArrayFull( length, value ) {
        let l_output = [ ];
        for ( let i = 0; i < length; i++ ) { l_output.push( value ); }
        return l_output;
    }
    function ArrayBroadCasting( l, l_bool, value ) {
        let l_output = [ ];
        for ( let i = 0; i < l.length; i++ ) { l_output.push( l_bool[ i ] ? value : l[ i ] ); }
        return l_output;
    }
    function ArrayCap( l, value ) { // cap value of a given list at the given capping value (manually set maximum value)
        l_modified = lodash.cloneDeep( l );
        for ( let i = 0; i < l.length; i ++ ) {
            if ( l[ i ] > value ) { // cap value of a given list at the given capping value (manually set maximum value)
                l_modified[ i ] = value;
            }
        }
        return l_modified
    }
    function ArrayFloor( l, value ) { // floor value of a given list at the given flooring value (manually set minimum value)
        l_modified = lodash.cloneDeep( l );
        for ( let i = 0; i < l.length; i ++ ) {
            if ( l[ i ] < value ) { // cap value of a given list at the given capping value (manually set maximum value)
                l_modified[ i ] = value;
            }
        }
        return l_modified
    }
    function ArrayApply( l, a_function ) { // apply certain function to all of values of a given array
        let l_output = [ ];
        for ( let i = 0; i < l.length; i ++ ) {
            l_output[ i ] = a_function( l[ i ] );
        }
        return l_output
    }
    function ArrayMap( l, dictionary ) { // map values in a given array using a given dictionary
        let l_output = [ ];
        for ( let i = 0; i < l.length; i ++ ) {
            l_output[ i ] = dictionary[ l[ i ] ];
        }
        return l_output
    }
    function ArrayTranspose( l_l ) { // transpose 2d array
        let n_row = l_l.length, n_col = l_l[ 0 ].length, l_l_t = [ ]; // transposed 2d array
        for ( let i = 0; i < n_col; i ++ ) {
            let l_t = [ ]; // transposed column (now a row)
            for ( let j = 0; j < n_row; j ++ ) { l_t.push( l_l[ j ][ i ] ); }
            l_l_t.push( l_t );
        }
        return l_l_t;
    }
    function ArrayAsInt( l ) { 
        let l_int = [ ];
        for ( let i = 0; i < l.length; i ++ ) {
            l_int.push( parseInt( l[ i ] ) ); // parse a string into an integer for each entry
        }
        return l_int
    } // convert array of strings to array of integers
    function ArrayAverage( l ) { // calculate average of values in the array
        let sum = l.reduce( ( a, b ) => a + b, 0 );
        let avg = ( sum / l.length ) || 0;
        return avg;
    }
    function ArrayMultiply( l, float_multiply_by ) { // return a list with values multiplied by 'float_multiply_by'
        let l_multiplied = [ ];
        for ( let i = 0; i < l.length; i ++ ) {
            l_multiplied.push( l[ i ] * float_multiply_by );
        }
        return l_multiplied;
    }
    function unpack( rows, key ) { 
        if ( Array.isArray( key ) ) { // if more than one key is given, return 2D-array of unpacked data
            let l_l_value = [ ];
            for ( let i = 0; i < rows.length; i ++ ) {
                let row = rows[ i ];
                let l_value = [ ];
                for ( let j = 0; j < key.length; j ++ ) { 
                    l_value.push( row[ key[ j ] ] );
                }
                l_l_value.push( l_value );
            }
            return l_l_value;
        } else { return rows.map( function( row ) { return row[ key ]; } ); } // if only a single key is given, return 1D-array of unpacked data
    };
    // tabular data
    function Unpack_String_rsa( seq ) { // unpack string containing RSA value using ASCII encoding (126 encodes null)
        let l_output = [ ];
        for ( let i = 0; i < seq.length; i ++ ) {
            let int_value = seq.charCodeAt( i ); // retrieve ASCII character as an integer number
            if ( int_value == 126 ) { l_output.push( null ); }
            else { l_output.push( Math.round( ( int_value - 33 ) / 92 * 100 ) / 100 ); } // RSA value range 0 ~ 1 // round up so that step is 1
        }
        return l_output;
    }
    function SplitColumn( dictarr, delimiter = '|' ) { // split column name of the given dict_array into two with the given delimiter character and return dict_dict_array # 20201106
        let dictdictarr = { };
        for ( let col in dictarr ) {
            let col_split = col.split( "|" );
            if ( ! ( col_split[ 0 ] in dictdictarr ) ) {
                dictdictarr[ col_split[ 0 ] ] = { };
            }
            dictdictarr[ col_split[ 0 ] ][ col_split[ 1 ] ] = dictarr[ col ];
        }
        return dictdictarr;
    }
    function SubsetDictDictArr( dictdictarr, l_col_inner, fill_empty_columm_if_not_exist = true, fill_value = 0 ) { // subset a given dictdictarr with a given list of l_col_inner list containing column names of the inner layer. if only one column is given, return dictarray containing the inner column # 20201106
        // 'fill_empty_columm_if_not_exist' : if true, fill empty column with 'fill_value' value
        if ( Array.isArray( l_col_inner ) ) { // subset a given dictdictarr with a given list of l_col_inner list containing column names of the inner layer.
            let dictdictarr_subset = { };
            for ( let col_outer in dictdictarr ) {
                for ( let i = 0; i < l_col_inner.length; i ++ ) {
                    let col_inner = l_col_inner[ i ];
                    if ( ! ( col_inner in dictdictarr[ col_outer ] ) ) {
                        if ( fill_empty_columm_if_not_exist ) { // 'fill_empty_columm_if_not_exist' : if true, fill empty column with 'fill_value' value
                            dictdictarr[ col_outer ][ col_inner ] = ArrayFull( dictdictarr[ col_outer ][ Object.keys( dictdictarr[ col_outer ] )[ 0 ] ].length, fill_value ); // assume that the length of array is same for all columns
                        }
                    } else {
                        if ( ! ( col_outer in dictdictarr_subset ) ) {
                            dictdictarr_subset[ col_outer ] = { };
                        }
                        dictdictarr_subset[ col_outer ][ col_inner ] = dictdictarr[ col_outer ][ col_inner ]
                    }
                }
            }
            return dictdictarr_subset;
        } else { // if only one column is given, return dictarray containing the inner column
            let col_inner_subset = l_col_inner;
            let dictarr = { };
            for ( let col_outer in dictdictarr ) {
                if ( ! ( col_inner_subset in dictdictarr[ col_outer ] ) ) {
                    dictarr[ col_outer ] = ArrayFull( dictdictarr[ col_outer ][ Object.keys( dictdictarr[ col_outer ] )[ 0 ] ].length, fill_value ); // assume that the length of array is same for all columns
                } else {
                    dictarr[ col_outer ] = dictdictarr[ col_outer ][ col_inner_subset ]
                }
            }
            return dictarr;
        }
    }
    function DictArrBooleanIndexing( dictarr, l_bool ) { // subset dictarr using a given boolean index (assumes the length of arrays for all keys are the same).
        let dictarr_new = { },
            l_key = Object.keys( dictarr );
        // for each array of a key in dictarr, perform boolean indexing using the given boolean index
        for ( let i = 0; i < l_key.length; i ++ ) { 
            let str_key = l_key[ i ];
            dictarr_new[ str_key ] = ArrayBooleanIndexing( dictarr[ str_key ], l_bool );
        }
        return dictarr_new;
    }
    function Pivot( l_index, l ) { // return dictarr where unique entries of l_index are keys and a list of correspondinb values to each entry (key) is a value to each key. // similar to 'Index_list_with_dictionary' function
        let dictarr = { };
        for ( let i = 0; i < l_index.length; i ++ ) {
            let index = l_index[ i ];
            if ( index in dictarr ) {
                dictarr[ index ].push( l[ i ] );
            } else {
                dictarr[ index ] = [ l[ i ] ];
            }
        }
        return dictarr;
    }
    // indexing
    function Index_list_with_dictionary( l ) { // build dictionary containing list for indices for a given list of accessions (key = accession, value = list of indices containing the accession)
        let dictionary_index = { };
        for ( let i = 0; i < l.length; i ++ ) { 
            let acc = l[ i ];
            if ( acc in dictionary_index ) {
                dictionary_index[ acc ].push( i )
            } else {
                dictionary_index[ acc ] = [ i ]; // initialize value of key encountered for the first time
            }
        }
        return dictionary_index;
    }
    // counter
    function Counter( l ) { // count entries in the list
        let dict_count = { };
        for ( entry of l ) {
            if ( entry in dict_count ) {
                dict_count[ entry ] += 1;
            } else {
                dict_count[ entry ] = 1;
            }
        }
        return dict_count;
    }
    function FindTheMajority( l ) { // count entries in the list and return the majority in the list
        let dict_count = Counter( l ), key_majority = '', count_majority = 0;
        for ( key of Object.keys( dict_count ) ) { // for each key
            let count = dict_count[ key ];
            if ( count > count_majority ) { // if the count of the current key is larger than count of majority key, update the majority key name
                key_majority = key;
                count_majority = count;
            }
        }
        return key_majority;
    }
    // colormap
    function Color_for_rsa( l_rsa ) {
        let l_color = [ ], function_colormap = function_colormap_rsa;
        for ( let i = 0; i < l_rsa.length; i ++ ) { l_color.push( l_rsa[ i ] == null? color_representing_null : function_colormap( l_rsa[ i ] * 2 ).toString( ) ); } // RSA of 0.5 will be represented as 'fully accessible to the surface'
        return l_color;
    }
    // select
    function Select_RemoveOptions( selectElement ) { // remove all option in the select element
       let length = selectElement.options.length - 1;
       for( let i = length; i >= 0; i -- ) {
          selectElement.remove( i );
       }
    }
    function Select_Get_Selected_Option_Using_ID( id_selectElement ) { // get selected option in the select element with its id
        let sel = document.getElementById( id_selectElement )
        return sel.options[ sel.selectedIndex ]
    }
    // specific to single cell expression data
    function RemoveNegativeCells( l_expr ) { // count and remove cells with zero counts as a expression value 
        let l_expr_without_negative_cells = [ ], n_cells_nagative = 0;
        for ( let i = 0; i < l_expr.length; i ++ ) {
            let expr = l_expr[ i ];
            if ( expr == 0 ) {
                n_cells_nagative += 1;
            } else {
                l_expr_without_negative_cells.push( expr );
            }
        }
        return { l_expr_without_negative_cells: l_expr_without_negative_cells, n_cells_nagative: n_cells_nagative };
    }
    // Load data
    function LoadRemoteTextFile( remoteFilePath ) { // load remote text file using xml_http_request (syncrhonous)
        let result = null;
        let xmlhttp = new XMLHttpRequest();
        xmlhttp.open( "GET", remoteFilePath, false );
        xmlhttp.send( );
        if ( xmlhttp.status == 200 ) {
            result = xmlhttp.responseText;
        }
        return result;
    }
    async function LoadLocalTextFile( file ) { // load and return a local text file using FileReader (asyncrhonous)
        return new Promise( ( resolve, reject ) => {
            let reader = new FileReader( ); // initialize FileReader
            reader.onload = function( progressEvent ) {
                console.log( "[LoadLocalTextFile] local text file", file, "loaded and parsed." );
                resolve( this.result ); // when loading and parsing is complete, resolve the promise
            };
            reader.readAsText( file )
        });
    }
    function LoadRemoteTextFile_PartialData( remoteFilePath, l_range = [ ] ) { // load remote text file using xml_http_request (syncrhonous) // l_range = list of ranges of bytes positions for partial downloads
        let result = null;
        let xmlhttp = new XMLHttpRequest( ),
            l_str_range = [ ];
        xmlhttp.open( "GET", remoteFilePath, false );
        if ( l_range.length == 0 ) { // handle empty l_range
            return '';
        } 
        xmlhttp.setRequestHeader('Range', 'bytes=' + l_range.map( l  => l.map( int_byte => int_byte.toString( ) ).join( '-' ) ).join( ', ' ) ); // compose a range header
        xmlhttp.send( );
        if ( xmlhttp.status == 206 ) {
            result = xmlhttp.responseText;
        }
        return result;
    }
    async function RetrieveRemoteFile_PartialData( l_index, dict_arr, dir_remote_file ) {
        // when 'flag_split_file' is true, assumes the main concatenated files are split into multiple file to satisfy maximum file size limits
        flag_split_file = 'index_file' in dict_arr; // check whether dict_arr contains the 'index_file' column
        if ( flag_split_file ) {
            let dict_index_file_to_l_range = { }, i = 0;
            // retrieve range containing records for each split file
            for ( i = 0; i < l_index.length ; i ++ ) {
                let index_record = l_index[ i ],
                    index_byte_start = dict_arr.index_byte_start[ index_record ],
                    index_byte_end = dict_arr.index_byte_end[ index_record ],
                    index_file = dict_arr.index_file[ index_record ];
                if ( ! ( index_file in dict_index_file_to_l_range ) ) {
                    dict_index_file_to_l_range[ index_file ] = [ ];
                }
                dict_index_file_to_l_range[ index_file ].push( [ index_byte_start + 1, index_byte_end - 1 ] ); // the first byte in each 'file' in the concatanated files is a spacer character (' ')
            }
            // retrieve data and parse data into records for each split file
            let l_record = [ ],
                l_index_file = Object.keys( dict_index_file_to_l_range );
            for ( i = 0; i < l_index_file.length; i ++ ) {
                // retrieve text records from a remote file
                let index_file = l_index_file[ i ],
                    l_range = dict_index_file_to_l_range[ index_file ];
                let text_combined = LoadRemoteTextFile_PartialData( dir_remote_file + '.' + index_file.toString( ), l_range ),
                    l_text = [ ];
                if ( l_range.length == 1 ) {
                    // when partial data of single part is retrieved, add all of the retrieved string to the list
                    l_text.push( text_combined );
                } else {
                    let l_split = text_combined.split( '\r\n' ).slice( 1, -1 ), index_text = 4; // retrieved data starts from the fifth line (index = 4)
                    for ( i = 0; i < l_range.length; i ++ ) {
                        l_text.push( l_split[ index_text ] ); 
                        index_text += 5; // retrieved data is in every fifth line
                    }
                }
                l_record.push( ... await ParseLocalData( l_text ) );
            }
            return l_record;
        } else {
            let l_range = [ ], i = 0;
            // retrieve text records from a remote file
            for ( i = 0; i < l_index.length ; i ++ ) {
                let index_record = l_index[ i ],
                    index_byte_start = dict_arr.index_byte_start[ index_record ],
                    index_byte_end = dict_arr.index_byte_end[ index_record ];
                l_range.push( [ index_byte_start + 1, index_byte_end - 1 ] ); // the first byte in each 'file' in the concatanated files is a spacer character (' ')
            }
            let text_combined = LoadRemoteTextFile_PartialData( dir_remote_file, l_range ),
                l_text = [ ];
            if ( l_range.length == 1 ) {
                // when partial data of single part is retrieved, add all of the retrieved string to the list
                l_text.push( text_combined );
            } else {
                let l_split = text_combined.split( '\r\n' ).slice( 1, -1 ), index_text = 4; // retrieved data starts from the fifth line (index = 4)
                for ( i = 0; i < l_range.length; i ++ ) {
                    l_text.push( l_split[ index_text ] ); 
                    index_text += 5; // retrieved data is in every fifth line
                }
            }
            return await ParseLocalData( l_text );
        }
    }
    async function ParseLocalData( l_text, flag_base64 = true ) { // parse local string or base64 string
        return new Promise( async function( resolve, reject ) { 
            let l_promise = l_text.map( function( text ) { 
                return new Promise ( function( resolve, reject ) {
                    // parse base64 strings
                    if ( flag_base64 ) {
                        text = ParseBase64GzippedText( text );
                    } 
                    Papa.parse( text, { download: false, header: false, skipEmptyLines: false, delimiter: '\t', worker: true, dynamicTyping: true, complete: function( results ) { resolve( ParseTransposedArrayWithHeader( results.data.slice( 0, results.data.length - 1 ) ) ) }, error: reject } ); // remove last empty line
                });
            }); // 'skipEmptyLines' is set to false to allow reading empty lines in some input files // due to 'skipEmptyLines: false' option, the last line is empty, and should be discarded
            let results = await Promise.all( l_promise ); // wait until all files have been parsed by Papa.parse
            let l_parsed_data = [ ]; 
            l_parsed_data.push( ... results );
            resolve( l_parsed_data );
        });
    }
    function ParseBase64GzippedText( str_base64_gzipped ) {
        // Decode base64 (convert ascii to binary) // Convert binary string to character-number array // Turn number array into byte-array // Convert gunzipped byteArray back to ascii string:
        let l_char = pako.inflate( new Uint8Array( atob( str_base64_gzipped ).split( '' ).map( function( x ) { return x.charCodeAt( 0 ); } ) ) ); 
        let str_content = ""; // convert ASCII value to character and concatenate the characters into a string
        for ( let i = 0; i < l_char.length; i ++ ) {
            str_content += String.fromCharCode( l_char[ i ] );
        }
        return str_content;
    }
    function ParseTransposedArrayWithHeader( l_l ) { // return a dictionary of arrays ( similar to the DataFrame.toDict() output ) from an array of arrays
        let dictarr = { };
        for ( let i = 0; i < l_l.length; i ++ ) {
            let key = l_l[ i ].shift( );
            dictarr[ key ] = l_l[ i ];
        }
        return dictarr;
    }
    async function LoadRemoteData( l_file ) { // load list of given tsv files from remote locations through asynchrous operations
        return new Promise( async function( resolve, reject ) { 
            let l_file_not_downloaded = [ ]; // list of files that were not downloaded
            for ( let i = 0; i < l_file.length; i ++ ) { 
                let url_file = l_file[ i ], name_file = url_file.slice( url_file.lastIndexOf( "/" ) + 1 );
                if ( ! ( name_file in object_data.file.dict_name_file_to_parsed_data ) ) { // if the name_file does not exist in the parsed data, add the url to 'l_file_not_downloaded'
                    l_file_not_downloaded.push( url_file );
                }
            }
            l_file = l_file_not_downloaded; // download files that are not already downloaded
            console.log( "[LoadRemoteData] remote data fetch started:", l_file );
            Loading_Screen( true ) // add loading screen
            let l_promise = l_file.map( function( file ) { 
                if( file.split( '.' ).slice( - 1 )[ 0 ] == "tsv" ) { // if given remote file is a tsv file
                    return new Promise ( function( resolve, reject ) {
                        Papa.parse( file, { download: true, header: false, skipEmptyLines: false, delimiter: '\t', worker: true, dynamicTyping: true, complete: function( results ) { resolve( ParseTransposedArrayWithHeader( results.data.slice( 0, results.data.length - 1 ) ) ) }, error: reject } ); // remove last empty line
                    });
                } else if ( file.split( '.' ).slice( - 2 ).join( '.' ) == "base64.txt" ) { // if given remote file is base64-encoded binary file
                    return new Promise ( function( resolve, reject ) {
                        fetch( file )
                            .then( ( response ) => {
                              return response.text( );
                           }).then( ( text_base64 ) => {
                              return ParseBase64GzippedText( text_base64 );
                           }).then( ( text ) => {
                              Papa.parse( text, { download: false, header: false, skipEmptyLines: false, delimiter: '\t', worker: true, dynamicTyping: true, complete: function( results ) { resolve( ParseTransposedArrayWithHeader( results.data.slice( 0, results.data.length - 1 ) ) ) }, error: reject } ); // remove last empty line
                           }).catch( err => { reject } );
                    });
                }
            }); // 'skipEmptyLines' is set to false to allow reading empty lines in some input files // due to 'skipEmptyLines: false' option, the last line is empty, and should be discarded
            let results = await Promise.all( l_promise ); // wait until all files have been parsed by Papa.parse
            let l_parsed_data = [ ]; 
            l_parsed_data.push( ... results );
            for ( i = 0; i < l_file.length; i ++ ) { 
                let url_file = l_file[ i ], name_file = url_file.slice( url_file.lastIndexOf( "/" ) + 1 );
                object_data.file.dict_name_file_to_parsed_data[ name_file ] = l_parsed_data[ i ];
            } // put parsed data into a dictionary
            console.log( "[LoadRemoteData] remote data loaded:", l_file );
            resolve( "[LoadRemoteData] resolved" )
            Loading_Screen( false ) // add loading screen
        });
    }
    async function ParseData( ) { // parse loaded data and prepare data objects
        return new Promise( async function( resolve, reject ) { 
            Loading_Screen( true );
            // initialize cache objects
            InitializeCacheForRemoteData( object_data.main.gene.expr, object_data.file.dict_name_file_to_parsed_data[ "expr.multiplied_by_1000.index.tsv.gz.base64.txt" ], [ 'gene_symbol' ], object_data.file.base_url + "expr.multiplied_by_1000.tsv.gz.base64.concatenated.txt" );
            InitializeCacheForRemoteData( object_data.main.cell_marker, object_data.file.dict_name_file_to_parsed_data[ "cell_marker.index.tsv.gz.base64.txt" ], [ 'tissue', 'cluster' ], object_data.file.base_url + "cell_marker.tsv.gz.base64.concatanated.txt" );
            let name_tissue = object_data.setting.app.name_tissue_default;
            object_data.main.gene.correl.dict_name_tissue_to_cache[ name_tissue ] = { }; // initialize the object
            InitializeCacheForRemoteData( object_data.main.gene.correl.dict_name_tissue_to_cache[ name_tissue ], object_data.file.dict_name_file_to_parsed_data[ `${object_data.main.gene.correl.dict_name_tissue_to_prefix[ name_tissue ]}.index.tsv.gz.base64.txt` ], [ 'gene_symbol' ], `${object_data.file.base_url}${object_data.main.gene.correl.dict_name_tissue_to_prefix[ name_tissue ]}.tsv.gz.base64.concatanated.txt` ); // initialize the cache for the correlation results of the current tissue
            // preview boxplot
            object_data.main.gene_summary = { };
            object_data.main.gene_summary.boxplot = { };
            object_data.main.gene_summary.boxplot.dictdictarr = SplitColumn( object_data.file.dict_name_file_to_parsed_data[ "all.box_precomputed_by_tissue.multiplied_by_1000.tsv.gz.base64.txt"] );

            console.log( "[ParseData] resolved" );
            Loading_Screen( false );
            resolve( "[ParseData] resolved" );
        });
    }
    // caching data from local or remote sources
    function InitializeCacheForRemoteData( object_cache, dict_arr_index, l_col_key, dir_remote_file, data_size = 5 ) {  // default data_size = 5
        object_cache.dir_remote_file = dir_remote_file;
        object_cache.dict_arr_index = dict_arr_index;
        // index 'dict_arr_index' with 'l_col_key', combining values from 'l_col_key' with '|'
        object_cache.l_col_key = l_col_key;
        let l_key = [ ], i = 0, j = 0;
        if ( l_col_key.length > 1 ) {
            let n_records = object_cache.dict_arr_index[ l_col_key[ 0 ] ].length;
            for ( i = 0; i < n_records; i ++ ) {
                let l_value = [ ];
                for ( j = 0; j < l_col_key.length; j ++ ) { 
                    l_value.push( dict_arr_index[ l_col_key[ j ] ][ i ] );
                }
                l_key.push( l_value.join( '|' ) );
            }
        } else {
            l_key = object_cache.dict_arr_index[ l_col_key[ 0 ] ];
        }
        object_cache.l_key = l_key;
        object_cache.dict_key_to_index = Index_list_with_dictionary( l_key );
        
        // add default attributes
        if ( ! ( 'data' in object_cache ) ) {
            object_cache[ 'data' ] = { };
        }
        if ( ! ( 'data_size' in object_cache ) ) {
            object_cache[ 'data_size' ] = data_size;
        }

    }
    function EvictCache( object_cache, l_key = null, flag_purge = false ) { // evict cache according to the amount of incoming data (l_key). Evict all data if flag_perge is true
        if ( flag_purge ) {
            // completely empty the cache data
            delete object_cache.data;
            object_cache.data = { }; 
            return true;
        }
//         console.log( "[EvictCache] evicting cache for the following incoming data", l_key );
        // retrieve list of keys that should be updated (keys that cannot be found in the cached data)
        let i = 0, j = 0,
            cache = object_cache.data,
            int_size_cache = object_cache.data_size,
            l_key_to_be_updated = [ ];
        for ( i = 0; i < l_key.length; i ++ ) {
            let str_key = l_key[ i ];
            if ( ! ( str_key in cache ) ) { // check cached data
                l_key_to_be_updated.push( str_key );
            }
        }
        // download data from a large indexed file, and add the new data to the cache
        if ( l_key_to_be_updated.length > int_size_cache ) {
            l_key_to_be_updated = l_key_to_be_updated.slice( 0, int_size_cache ); // slice the list of requested genes so that the number of retrieved gene does not exceed that of cache size.
//             console.log( "[EvictCache] number of incoming records exceeds the maximum size of cache, only retrieving records in the front. The following keys will be fetched", l_key_to_be_updated );
        }
        let int_n_records_newly_added = l_key_to_be_updated.length,
            int_n_records_to_be_evicted = Math.max( 0, Object.keys( cache ).length + int_n_records_newly_added - int_size_cache );
        // iteratively evict 'int_n_records_to_be_evicted' number of records based on the last used time
        for ( i = 0; i < int_n_records_to_be_evicted; i ++ ) {
            let l_key_in_a_cache = Object.keys( cache ), // one key is used to store the size of cache
                str_key_least_recently_used = -1,
                time_least_recently_used = -1;
            for ( j = 0; j < l_key_in_a_cache.length; j ++ ) {
                let str_key = l_key_in_a_cache[ j ],
                    time_last_used = cache[ str_key ].last_used_time;
                if ( ( time_least_recently_used == -1 ) | ( time_last_used < time_least_recently_used ) ) {
                    str_key_least_recently_used = str_key;
                    time_least_recently_used = time_last_used;
                }
            }
//             console.log( "[EvictCache] evicting", str_key_least_recently_used );
            delete cache[ str_key_least_recently_used ];
        }
        return l_key_to_be_updated;
    }
    async function UpdateCache( object_cache, function_preprocessing_data = null, l_key_to_be_updated_from_remote_source = null, dict_incoming_data = null ) { // update cache using either locally available data 'dict_incoming_data' (key = key for caching, value = incoming data) or remotely available data 'l_key_to_be_updated_from_remote_source'
        return new Promise( async function( resolve, reject ) { 
            let i = 0, flag_cache_evicted = false; // a flag indicating whether cache has been evicted 
            if ( dict_incoming_data === null ) {
                if ( l_key_to_be_updated_from_remote_source === null ) {
                    console.log( '[UpdateCache] required arguments not given' );
                    return false;
                }
                l_key_to_be_updated_from_remote_source = EvictCache( object_cache, l_key_to_be_updated_from_remote_source ); // evict (free) cache data before downloading new data from the remote source
                flag_cache_evicted = true;
                let l_index_key = [ ];
                for ( i = 0; i < l_key_to_be_updated_from_remote_source.length; i ++ ) {
                    l_index_key.push( object_cache.dict_key_to_index[ l_key_to_be_updated_from_remote_source[ i ] ][ 0 ] );
                }
                let l_record = await RetrieveRemoteFile_PartialData( l_index_key, object_cache.dict_arr_index, object_cache.dir_remote_file );
                dict_incoming_data = { };
                for ( i = 0; i < l_record.length; i ++ ) {
                    dict_incoming_data[ l_key_to_be_updated_from_remote_source[ i ] ] = l_record[ i ];
                }
            }
            let l_key = Object.keys( dict_incoming_data );
            if ( ! flag_cache_evicted ) {
                l_key = EvictCache( object_cache, l_key ); // evict (free) cache data before adding new data to the cache
            }
            for ( i = 0; i < l_key.length; i ++ ) {
                let str_key = l_key[ i ],
                    record = dict_incoming_data[ str_key ];
                object_cache.data[ str_key ] = { data : ( function_preprocessing_data === null ) ? record : function_preprocessing_data( record ), last_used_time : new Date( ).getTime( ) }; // add new data to the cache
            }
    //         console.log( '[UpdateCache] cache data updated with the following data:', dict_incoming_data );
            resolve( "[UpdateCache] Resolved" ); 
        });
    }
    
    var div_current_view = $( "#application-current-view" ); // global variable
    function DownloadCurrentView( ) {
         html2canvas( div_current_view.get( 0 ) ).then ( function ( canvas ) {
                var imgageData = canvas.toDataURL("image/png");
                // Now browser starts downloading it instead of just showing it
                var newData = imgageData.replace(/^data:image\/png/, "data:application/octet-stream");
                $("#anchor-current-view").attr("download", "Single Cell long-noncoding RNA Atlas-screenshot.png").attr( "href", newData );
                $("#anchor-current-view").get( 0 ).click( ) // download image file
         });
    }
    
    var searchbox_main = document.getElementById( 'searchbox_main' ), tagify_searchbox_main = { };
    object_data.flags.flag_tagify_searchbox_main_mutationobserver_attached = false;
    function Initialize_SearchBox_Main( ) {
        tagify_searchbox_main = new Tagify( searchbox_main, {
            dropdown: { 
                maxItems: parseInt( ( window.innerHeight > 1500 ) ? 100 : Math.max( 1, ( window.innerHeight - 500 ) / 100 ) ), // set max number of dropdown items according to the window height
                enabled: 0,
                position: 'text', // enable minimal sizing of a dropdown list
                highlightFirst: true, // enable enhanced autocomplete feature
            },
            maxTags: object_data.main.gene.expr.data_size, // max number of tags = max size of gene expression data cache
            editTags: false, // disable editing
            enforceWhitelist: true, whitelist: object_data.file.dict_name_file_to_parsed_data[ "expr.multiplied_by_1000.index.tsv.gz.base64.txt" ][ "gene_symbol" ]
        })
        // Chainable event listeners
        tagify_searchbox_main
              .on( 'add', e => { 
            console.log( "onAddTag: ", e.detail.data.value );
            if ( object_data.main.scatter.flag_waiting_selection_of_gene ) { // use the selected gene as one of the axes of the scatter plot
                object_data.main.scatter.gene_symbol_selected = e.detail.data.value;
                Change_UMAP_Coordinate( null );
            } else { // use the selected gene to color points
                object_data.main.gene.gene_symbol = e.detail.data.value;
                UpdateGene( flag_restyle = true );
            }
            $( e.detail.tag ).mouseenter( function( e ) {
                console.log( 'mouse enter', e.target.textContent )
                if ( object_data.setting.plot.preview_gene.flag_preview_on_hover ) {
                    let gene_symbol = e.target.textContent; // retrieve gene_symbol
                    // preview a gene on hover
                    $( "#container_plotly_searchbox_main_preview" ).collapse( 'show' ); // show boxplot panel
                    PreviewGene( gene_symbol );
                }
            })
                .mouseleave( function( e ) {
                console.log( 'mouse leave', e.target.textContent )
                // hide boxplot panel
                $( "#container_plotly_searchbox_main_preview" ).collapse( 'hide' ); // show boxplot panel
            })
                .mousemove( function( e ) {
                console.log( 'mouse move', e.target.textContent )
                if ( object_data.setting.plot.preview_gene.flag_preview_on_hover ) {
                    // show boxplot panel
                    $( "#container_plotly_searchbox_main_preview" ).collapse( 'show' ); // show boxplot panel
                }
            })
        } )
              .on( 'click', e => { 
            console.log( "onClickTag: ", e.detail.data.value );
            if ( object_data.main.scatter.flag_waiting_selection_of_gene ) { // use the selected gene as one of the axes of the scatter plot
                object_data.main.scatter.gene_symbol_selected = e.detail.data.value;
                Change_UMAP_Coordinate( null );
            } else { // use the selected gene to color points
                object_data.main.gene.gene_symbol = e.detail.data.value;
                UpdateGene( flag_restyle = true );
            }
        } )
              .on( 'remove', e => { console.log( "onRemoveTag: ", e.detail ); } )
              .on('input', e => { // preview during typing
            if ( ! isString( e.detail.value ) ) { return -1; }
            let gene_symbol = e.detail.value; // retrieve current input
            if ( gene_symbol in object_data.main.gene.expr.dict_key_to_index ) { // show preview of the gene if it exists in the data
                PreviewGene( gene_symbol );
            }
        } )
              .on( "dropdown:hide", ( e ) => { console.log( "dropdown:hide" ); $( "#container_plotly_searchbox_main_preview" ).collapse( 'hide' ); } )
              .on( "dropdown:show", ( e ) => { // preview during exploring dropdown options
            $( "#container_plotly_searchbox_main_preview" ).collapse( 'show' );
            console.log( "dropdown:show" );
            if ( ! object_data.flags.flag_tagify_searchbox_main_mutationobserver_attached ) {
                let dropdown = document.querySelector( '.tagify__dropdown' ); // retrieve DOM of active dropdown
                let options = {
                    subtree: true,
                    attributes: true
                };
                let observer = new MutationObserver( function( mutations ) {
                    mutations.forEach( function( e ) {
                        if ( ( e.attributeName == 'class' ) & ( e.target.className == "tagify__dropdown__item tagify__dropdown__item--active" ) ) { // detect the innerHTML of the active element using MutationObserver (the closely matched gene_symbol)
                            console.log( "dropdown active element:", e.target.innerHTML );
                            PreviewGene( e.target.innerHTML );
                        }
                    });
                });
                observer.observe( dropdown, options );
                let dragsort = new DragSort( tagify_searchbox_main.DOM.scope, {
                    selector: '.' + tagify_searchbox_main.settings.classNames.tag,
                    callbacks: {
                        dragEnd: ( elm ) => { tagify_searchbox_main.updateValueByDOMTags( ); }
                    }
                });
                object_data.flags.flag_tagify_searchbox_main_mutationobserver_attached = true; // set the flag to true after attaching mutationobserver to the dropdown list so that another observer is not attached.
            }
        })
    }

    async function Change_UMAP_Coordinate( str_tissue = null ) { 
        if ( str_tissue === null ) {
            // change x or y-axis of scatter plot using the current setting
            let select_x_axis = $id( 'select_scatter_x_axis' ),
                select_y_axis = $id( 'select_scatter_y_axis' ),
                x_axis_option_current = select_x_axis.value,
                y_axis_option_current = select_y_axis.value,
                x_axis_option_previous = object_data.main.scatter.x_axis_option,
                y_axis_option_previous = object_data.main.scatter.y_axis_option,
                flag_update_x_axis = x_axis_option_current != x_axis_option_previous,
                flag_update_y_axis = y_axis_option_current != y_axis_option_previous,
                flag_update_axis = flag_update_x_axis | flag_update_y_axis;
            if ( ! flag_update_axis ) {
                console.log( "[Change_UMAP_Coordinate] no change in x- or y-axis setting detected, exiting" );
                return -1
            }
            if ( flag_update_x_axis ) { // update x-axis
                let option_gene = select_x_axis.options[ 1 ];
                if( x_axis_option_current == '__UMAP__' ) {
                    object_data.main.scatter.x_label = 'UMAP-1';
                    object_data.main.scatter.x_axis_option = x_axis_option_current;
                    object_data.main.scatter.x_axis_col_name = '__UMAP__';
                    // reset select option
                    option_gene.value = 'Gene';
                    option_gene.innerHTML = 'Gene';
                    object_data.main.scatter.flag_axis_using_gene_expr = false;
                } else if ( object_data.main.scatter.flag_waiting_selection_of_gene ) {
                    // update setting for scatter plot
                    let gene_symbol = object_data.main.scatter.gene_symbol_selected;
                    object_data.main.scatter.flag_waiting_selection_of_gene = false; // reset flag
                    object_data.main.scatter.x_label = gene_symbol;
                    object_data.main.scatter.x_axis_option = gene_symbol;
                    object_data.main.scatter.x_axis_col_name = '__x_axis_gene_expr__';
                    object_data.main.scatter.flag_axis_using_gene_expr = true;
                    // update select option
                    option_gene.value = gene_symbol;
                    option_gene.innerHTML = gene_symbol;
                } else {
                    object_data.main.scatter.flag_waiting_selection_of_gene = true;
                    // lock selection button to allow the selection of axis by clicking gene_symbol
                    select_x_axis.disabled = true;
                    select_y_axis.disabled = true;
                    console.log( '[Change_UMAP_Coordinate] (X-axis) waiting a gene_symbol to be selected' );
                    return 1;
                }
            } else { // update y-axis
                let option_gene = select_y_axis.options[ 1 ];
                if( y_axis_option_current == '__UMAP__' ) {
                    object_data.main.scatter.y_label = 'UMAP-2';
                    object_data.main.scatter.y_axis_option = y_axis_option_current;
                    object_data.main.scatter.y_axis_col_name = '__UMAP__';
                    // reset select option
                    option_gene.value = 'Gene';
                    option_gene.innerHTML = 'Gene';
                    object_data.main.scatter.flag_axis_using_gene_expr = false;
                } else if ( object_data.main.scatter.flag_waiting_selection_of_gene ) {
                    // update setting for scatter plot
                    let gene_symbol = object_data.main.scatter.gene_symbol_selected;
                    object_data.main.scatter.flag_waiting_selection_of_gene = false; // reset flag
                    object_data.main.scatter.y_label = gene_symbol;
                    object_data.main.scatter.y_axis_option = gene_symbol;
                    object_data.main.scatter.y_axis_col_name = '__y_axis_gene_expr__';
                    object_data.main.scatter.flag_axis_using_gene_expr = true;
                    // update select option
                    option_gene.value = gene_symbol;
                    option_gene.innerHTML = gene_symbol;
                } else {
                    object_data.main.scatter.flag_waiting_selection_of_gene = true;
                    // lock selection button to allow the selection of axis by clicking gene_symbol
                    select_x_axis.disabled = true;
                    select_y_axis.disabled = true;
                    console.log( '[Change_UMAP_Coordinate] (Y-axis) waiting a gene_symbol to be selected' );
                    return 1;
                }
            }
        } else { 
            // change UMAP coordinates to the UMAP coordinates of the given tissue
            if ( str_tissue == object_data.main.umap.tissue ) {
                console.log( "[Change_UMAP_Coordinate] UMAP coordinates of", str_tissue, "already drawn, exiting" );
                return -1;
            }
            object_data.main.umap.tissue = str_tissue; // update str_tissue of the main UMAP graph
            object_data.main.umap.flag_combined_tissues = str_tissue in object_data.main.combined; // flag indicating whether UMAP coordinates of combined tissues will be used.
            let dictarr = object_data.file.dict_name_file_to_parsed_data[ "cells.all.tsv.gz.base64.txt" ]; // retrieve data
            
            if ( ! ( str_tissue in object_data.main.gene.correl.dict_name_tissue_to_cache ) ) { // correlation result cache for the name_tissue is not available
                // initialize correlation result cache for the name_tissue
                object_data.main.gene.correl.dict_name_tissue_to_cache[ str_tissue ] = { }; // initialize the object
                await LoadRemoteData( [ `${object_data.file.base_url}${object_data.main.gene.correl.dict_name_tissue_to_prefix[ str_tissue ]}.index.tsv.gz.base64.txt` ] ); // load index data from remote locations (lazy loading)
                InitializeCacheForRemoteData( object_data.main.gene.correl.dict_name_tissue_to_cache[ str_tissue ], object_data.file.dict_name_file_to_parsed_data[ `${object_data.main.gene.correl.dict_name_tissue_to_prefix[ str_tissue ]}.index.tsv.gz.base64.txt` ], [ 'gene_symbol' ], `${object_data.file.base_url}${object_data.main.gene.correl.dict_name_tissue_to_prefix[ str_tissue ]}.tsv.gz.base64.concatanated.txt` ); // initialize the cache for the correlation results of the current tissue
            }
            
            if ( ! ( str_tissue in object_data.main.filters_for_subsets ) ) { // if the boolean filter does not exist, build the filter
                // build a boolean index for retrieving UMAP coordinates (and other data) for the current tissue
                let l_tissue = dictarr[ "tissue" ],
                    l_bool_index = [ ],
                    i = 0;
                if ( object_data.main.umap.flag_combined_tissues ) {
                    let tissues_of_a_combined_dataset = object_data.main.combined[ str_tissue ]; // retrieve name of the tissues of the combined dataset
                    for ( i = 0; i < l_tissue.length; i ++ ) {
                        l_bool_index.push( l_tissue[ i ] in tissues_of_a_combined_dataset ); // True if the tissue exists in the list
                    }
                } else {
                    for ( i = 0; i < l_tissue.length; i ++ ) {
                        l_bool_index.push( l_tissue[ i ] == str_tissue ); // True if the tissue match the current tissue
                    }
                }
                object_data.main.filters_for_subsets[ str_tissue ] = l_bool_index; // save the composed filter
            }
            
            let l_bool_index = object_data.main.filters_for_subsets[ str_tissue ]; // retrieve filters
            object_data.main.umap.l_bool_index_current_cells = l_bool_index; // set the filter
            object_data.main.umap.dictarr = DictArrBooleanIndexing( dictarr, l_bool_index ); // subset the data
        }
        // Update coordinates of scatter plots
        // draw cell annotation plot
        PlotCellAnnotation( );
        // draw expression plot
        UpdateGene( );
    }
    // preprocessing functions for updating caches
    function preprocess_expr( data ) {
        let str_key = Object.keys( data )[ 0 ];
        return ArrayMultiply( data[ str_key ], Math.pow( 0.1, object_data.setting.expression.n_digits_below_floating_point ) );
    }
    var plotPreviewGene = document.getElementById( "plotly_searchbox_main_preview" ),
        plotSingleCellAnnotation = document.getElementById( "plotly_sclncrna_expression_umap_annotation" ),
        plotSingleCellExpression = document.getElementById( "plotly_sclncrna_expression_umap_expression" ),
        plotSingleCellExpressionViolin = document.getElementById( "plotly_sclncrna_expression_violine" );
    object_data.setting.plot.single_cell_scatter = { 
        max_height: 750,
        marker_size: 2,
        int_factor_axis_range_to_marker_size : 30, 
        int_factor_marker_size_multiply : 1
    }; // defsult single-cell scatter plot setting
    object_data.setting.plot.single_cell_violin = { };
    var config_plotly = { responsive: true, displaylogo: false, toImageButtonOptions: { hoverClosestCartesian: true, format: 'png', filename: name_application, height: 750, width: 1100, scale: 1 }, modeBarButtonsToRemove: [ 'autoScale2d', 'toggleSpikelines', 'hoverCompareCartesian', 'hoverClosestCartesian' ] } // flexible plot width according to the window size  // do not display plotly logo
    function SortViolinTrace_Expression( trace_a, trace_b ) { return - ( trace_a.mean_expression_for_sorting - trace_b.mean_expression_for_sorting ) } // Sort Plotly Violin traces displaying expression values based on parameter values
// assign color to violine plot traces
    function String_Check_Substring( string, l_substring ) {
        let flag_contains_substring = false;
        string = string.toLowerCase( );
        for ( let i = 0; i < l_substring.length; i ++ ) {
            let substring = l_substring[ i ].toLowerCase( );
            if ( string.includes( substring ) ) {
                flag_contains_substring = true;
                break;
            }
        }
        return flag_contains_substring;
    }
    function PlotExpression_scRNA_Seq_Get_Color_for_name_clus( str_name_clus ) { // return color for the given str_name_clus
        str_name_clus = str_name_clus.toLowerCase( ); // convert name_clus to lower cases
        let color = "#48D1CC"
        if ( String_Check_Substring( str_name_clus, [ 'platelet' ] ) ) {
            color = "#aec2c1";
        } else if ( String_Check_Substring( str_name_clus, [ 'mast cell' ] ) ) {
            color = "#ceb3fc";
        } else if ( String_Check_Substring( str_name_clus, [ 'uninfected' ] ) ) {
            color = "#9ef274";
        } else if ( String_Check_Substring( str_name_clus, [ 'infected' ] ) ) {
            color = "#ff6b6b";
        } else if ( String_Check_Substring( str_name_clus, [ 'rbc' ] ) ) {
            color = "#fc4970";
        } else if ( String_Check_Substring( str_name_clus, [ 'endo', 'lymphatic', 'endotheli', 'capillary', 'luminal' ] ) ) {
            color = "#faf75f";
        } else if ( String_Check_Substring( str_name_clus, [ 'dendritic', 'pdc', 'cdc', 'dc_', ' dcs' ] ) ) {
            color = "#64b9fa";
        } else if ( String_Check_Substring( str_name_clus, [ 'stromal', 'fibroblasts', 'fibs', 'mesenchyme', 'cartilage' ] ) ) {
            color = "#fa5f5f";
        } else if ( String_Check_Substring( str_name_clus, [ 'goblet', 'ciliated', 'secretory', 'ionocyte', 'club' ] ) ) {
            color = "#fce6b6";
        } else if ( String_Check_Substring( str_name_clus, [ 'alveolar', 'basal', 'squamous', 'epithelia' ] ) ) {
            color = "#ffbad9";
        } else if ( String_Check_Substring( str_name_clus, [ 'tcell', 't cell', 't_cell', 't_', 'gd t', 'cd8', 'cd4', 'treg' ] ) ) {
            color = "#b89eff";
        } else if ( String_Check_Substring( str_name_clus, [ 'mono', 'lymphoid', 'hematopo', 'leukocyt', 'immuno', 'monocyte', 'neutrop', 'macro', 'b_cell', 'b cell', 'plasma cell', 'nk', 'eosinophil', 'naive' ] ) ) {
            color = "#e79eff";
        } else if ( String_Check_Substring( str_name_clus, [ 'myo' ] ) ) {
            color = "#e8a78b";
        }
        return color;
    }
    async function UpdateGene( flag_restyle = false ) { // draw a new plot by default
        console.log( "[UpdateGene] updating gene started" );
        let str_tissue = object_data.main.umap.tissue, // retrieve current tissue
            gene_symbol = object_data.main.gene.gene_symbol;
        
        // update plot (Single Cell Expression)
        // update cache of expression data
        let object_cache = object_data.main.gene.expr;
        await UpdateCache( object_cache, preprocess_expr, [ gene_symbol ] ); // retrieve the single cells expression data of the current gene.

        plotSingleCellExpression.style.height = Math.min( plotSingleCellExpression.offsetWidth, object_data.setting.plot.single_cell_scatter.max_height ) + 'px'; // responsive height ( width = height )
    
        let dictarr = object_data.main.umap.dictarr,
            l_expr = object_data.main.gene.expr.data[ gene_symbol ].data,
            col_name_cell_type = ( object_data.main.umap.flag_combined_tissues ) ? 'cell_type' : 'tissue_cell_type';
        // retrieve expression values
        l_expr = ArrayBooleanIndexing( l_expr, object_data.main.umap.l_bool_index_current_cells );
        
        // update x-axis values from gene expression data
        if ( object_data.main.scatter.x_axis_col_name == '__x_axis_gene_expr__' ) {
            let gene_symbol_x_axis = object_data.main.scatter.x_axis_option,
                l_expr_x_axis = [ ];
            await UpdateCache( object_cache, preprocess_expr, [ gene_symbol_x_axis ] ); // retrieve the single cells expression data of the current gene.
            l_expr_x_axis = ArrayBooleanIndexing( object_data.main.gene.expr.data[ gene_symbol_x_axis ].data, object_data.main.umap.l_bool_index_current_cells ); // retrieve x-axis
            dictarr[ '__x_axis_gene_expr__' ] = l_expr_x_axis;
        } else {
            object_data.main.scatter.x_axis_col_name = ( object_data.main.umap.flag_combined_tissues ) ? 'UMAP_1' : 'tissue_UMAP_1';
        }
        if ( object_data.main.scatter.y_axis_col_name == '__y_axis_gene_expr__' ) {
            let gene_symbol_y_axis = object_data.main.scatter.y_axis_option,
                l_expr_y_axis = [ ];
            await UpdateCache( object_cache, preprocess_expr, [ gene_symbol_y_axis ] ); // retrieve the single cells expression data of the current gene.
            l_expr_y_axis = ArrayBooleanIndexing( object_data.main.gene.expr.data[ gene_symbol_y_axis ].data, object_data.main.umap.l_bool_index_current_cells ); // retrieve y-axis
            dictarr[ '__y_axis_gene_expr__' ] = l_expr_y_axis;
        } else {
            object_data.main.scatter.y_axis_col_name = ( object_data.main.umap.flag_combined_tissues ) ? 'UMAP_2' : 'tissue_UMAP_2';
        }
        // complete the selection process by re-enabling the selection button
        $id( 'select_scatter_x_axis' ).disabled = false; 
        $id( 'select_scatter_y_axis' ).disabled = false; 
        object_data.setting.plot.single_cell_scatter.int_factor_axis_range_to_marker_size = object_data.main.scatter.flag_axis_using_gene_expr ? 6 : 30;
        object_data.setting.plot.single_cell_scatter.marker_size = ResponsiveMarkerSize( plotSingleCellExpression, object_data.setting.plot.single_cell_scatter.int_factor_axis_range_to_marker_size, object_data.setting.plot.single_cell_scatter.int_factor_marker_size_multiply );
        console.log( `marker size is ${object_data.setting.plot.single_cell_scatter.marker_size}` );
        if ( flag_restyle ) { // restyle existing plot
            console.log( "[UpdateGene] restyling existing plot for", gene_symbol );
            let update = { 
                hovertemplate: ["Expression of <b>" + gene_symbol + "</b>: <b>%{marker.color:.3f}</b>",
                                "tissue: <b>%{customdata[1]}</b>",
                                "cluster: <b>%{customdata[0]}</b>",
                                "",
                                object_data.main.scatter.x_label + ": <b>%{x:.2f}</b>",
                                object_data.main.scatter.y_label + ": <b>%{y:.2f}</b>",
                                "%{customdata[2]}",
                                "<extra></extra>" ].join( "<br>" ),
                marker: { 
                    color: l_expr,
                    showscale: true,
                    colorscale: "YlOrRd",
                    reversescale: true,
                    size: object_data.setting.plot.single_cell_scatter.marker_size
                } };
            Plotly.restyle( plotSingleCellExpression, update, [ 0 ] );
        } 
        else { // draw a new plot
            let trace = {
                mode: 'markers',
                type: 'scattergl',
                x: dictarr[ object_data.main.scatter.x_axis_col_name ],
                y: dictarr[ object_data.main.scatter.y_axis_col_name ],
                customdata: ArrayTranspose( [ dictarr[ col_name_cell_type ],
                                              dictarr[ "tissue" ],
                                              dictarr[ 'id_cell' ]
                                            ] ),
                hovertemplate: ["Expression of <b>" + gene_symbol + "</b>: <b>%{marker.color:.3f}</b>",
                                "tissue: <b>%{customdata[1]}</b>",
                                "cell type: <b>%{customdata[0]}</b>",
                                "",
                                object_data.main.scatter.x_label + ": <b>%{x:.2f}</b>",
                                object_data.main.scatter.y_label + ": <b>%{y:.2f}</b>",
                                "%{customdata[2]}",
                                "<extra></extra>" ].join( "<br>" ),
                marker: { 
                    color: l_expr,
                    showscale: true,
                    colorscale: "YlOrRd",
                    reversescale: true,
                    size: object_data.setting.plot.single_cell_scatter.marker_size
                },
                line: { 
                    width: 0
                },
            };
            Plotly.newPlot( plotSingleCellExpression, [ trace ], { margin: { l: 65, r: 65, b: 90, t: 50, pad: 4 }, xaxis: { title : object_data.main.scatter.x_label, zeroline: false, showgrid : false }, yaxis: { title : object_data.main.scatter.y_label, zeroline: false, showgrid : false }, hovermode: "closest" }, { ... config_plotly, filename: name_application + '__Expression_of_' + gene_symbol, height: 1100, width: 1100 } ); // draw scatter plot of the annotation
            plotSingleCellExpression.on( "plotly_relayout" , function( ed ) {
                PlotlySynchronizedLayout( ed, l_div_plotly );
            });
            plotSingleCellExpression.on( 'plotly_click', async function( data ) {
                PlotlyCallBackSingleCellMarker( data.points[ 0 ] );
            });
        }
        // Update correlated genes 
        
        object_cache = object_data.main.gene.correl.dict_name_tissue_to_cache[ object_data.main.umap.tissue ]; // retrieve the cache of correlation result of the current tissue
        await UpdateCache( object_cache, null, [ gene_symbol ] ); // update cache of correlated gene data
        Display_DataTable( 'container_datatable_correl', 'datatable_correl', object_cache.data[ gene_symbol ].data, [ "gene_symbol", "correlation_coefficient", "probability" ] );
        // change panel name
        document.getElementById( "panel_name_correl_gene").innerHTML = `Co-Expressed Genes of ${gene_symbol} in "${object_data.main.umap.tissue}"`;
        // change datatable name 
        object_data.main.datatable[ 'datatable_correl' ].name = `correl_gene.${gene_symbol}.${object_data.main.umap.tissue}`;
        // show datatable panel
        $( "#container_panel_datatable_correl" ).collapse( 'show' )
        
        // Update interaction graph
//         LoadRemoteTextFile( "http://www.rna-society.org/rnainter/php_mysql/api.php?keyword=Actb&species=Mus musculus" )
        
        // Violin plots
        object_data.setting.plot.single_cell_violin.ignore_cells_with_zero_counts = ! document.getElementById( "checkbox__plotly_sclncrna_expression_violine__ignore_cells_with_zero_counts" ).checked; // retrieve current setting
        // build list of traces for each category
        dictarr = Pivot( object_data.main.umap.dictarr[ col_name_cell_type ], l_expr ); // key = cluster, value = list of expression values.
        let trace_template = { 
            type: 'violin',
            opacity: 0.6,
            box: { visible: true, width: 0.075 },
            meanline: { visible: true }
        },
            l_category = Object.keys( dictarr ),
            str_separation = ( plotSingleCellExpressionViolin.offsetWidth > 3000 ) ? '<br>' : ' ' ; // set 'str_separation' based on current width to increase responsiveness of the plot
        let l_trace = [ ];
        for ( let category in dictarr ) {
            let l_expr = dictarr[ category ],
                n_cells = l_expr.length, // retrieve number of cells for the current category
                float_mean_expression = ArrayAverage( l_expr ), // retrieve a mean expression of cells for the current trace
                object_RemoveNegativeCells = RemoveNegativeCells( l_expr ), // retrieve list of expression values excluding cells with zero counts
                n_cells_with_zero_count = object_RemoveNegativeCells.n_cells_nagative,
                l_expr_without_negative_cells = object_RemoveNegativeCells.l_expr_without_negative_cells;

            let trace = { ... trace_template,
                         x0: "", 
                         y: [ ],
                         name: category,
                         line: { color: 'black', width: 2 },
                         fillcolor: object_data.setting.plot.annotation.dict_tissue_to_dict_cell_type_to_color[ str_tissue ][ category ],
                         points: false, // do not show outliers to increase responsivity
                         y_ignore_cells_with_zero_counts__true: l_expr_without_negative_cells,
                         y_ignore_cells_with_zero_counts__false: ArrayFull( n_cells_with_zero_count, 0 ).concat( l_expr_without_negative_cells ),
                         x0_ignore_cells_with_zero_counts__true: category + str_separation + "(n=" + ( n_cells - n_cells_with_zero_count ) + ' of ' + n_cells + ')',
                         x0_ignore_cells_with_zero_counts__false: category + str_separation + "(n=" + n_cells + ')',
                         mean_expression_for_sorting: float_mean_expression
                        };
            l_trace.push( trace );
        }
        l_trace = l_trace.sort( SortViolinTrace_Expression ); // sort by mean expression
        let layout = {
            margin: { l: 50, r: 65, b: 150, t: 50, pad: 4 },
            yaxis: { title : "Normalized Expression", rangemode: "nonnegative" },
            legend: { tracegroupgap: 0 },
            violingap: 0, // split violin plot for two covid statuses
            violingroupgap: 0,
            violinmode: "overlay",
        }
        for ( let i = 0; i < l_trace.length; i ++ ) { // swap y and x0 values based on the 'object_plot_scrnaseq.option.ignore_cells_with_zero_counts' setting
            l_trace[ i ].x0 = object_data.setting.plot.single_cell_violin.ignore_cells_with_zero_counts ? l_trace[ i ].x0_ignore_cells_with_zero_counts__true : l_trace[ i ].x0_ignore_cells_with_zero_counts__false ;
            l_trace[ i ].y = object_data.setting.plot.single_cell_violin.ignore_cells_with_zero_counts ? l_trace[ i ].y_ignore_cells_with_zero_counts__true : l_trace[ i ].y_ignore_cells_with_zero_counts__false ;
        }
        Plotly.newPlot( plotSingleCellExpressionViolin, l_trace, layout, { ... config_plotly, filename: name_application + '__Expression_of_' + gene_symbol, height: 600, width: 1100 } );
    }

    async function PlotCellAnnotation( flag_restyle = false ) { // draw a new plot by default
        let str_tissue = object_data.main.umap.tissue, // retrieve current tissue
            option_selected = Select_Get_Selected_Option_Using_ID( "select_umap_annotation_data" ),
            str_name_data_annotation = option_selected.value, // retrieve name of data for annotations
            str_name_option_selected = option_selected.innerHTML; // retrieve the name of the selected option
        
        plotSingleCellAnnotation.style.height = Math.min( plotSingleCellAnnotation.offsetWidth, object_data.setting.plot.single_cell_scatter.max_height ) + 'px'; // responsive height ( width = height )
        // demonstration
        let object_cache = object_data.main.gene.expr;
        let dictarr = object_data.main.umap.dictarr, 
            col_name_cell_type = ( object_data.main.umap.flag_combined_tissues ) ? 'cell_type' : 'tissue_cell_type',
            l_color = [ ], marker_colormap = { }, str_hovertemplate_additional_line = '';
        // set color 
        switch( str_name_data_annotation ) {
          case 'tissue':
            l_color = ArrayMap( dictarr[ str_name_data_annotation ], object_data.setting.plot.annotation.dict_tissue_to_color );
            break;
          case 'cell type':
            l_color = ArrayMap( dictarr[ col_name_cell_type ], object_data.setting.plot.annotation.dict_tissue_to_dict_cell_type_to_color[ str_tissue ] );
            break;
          default:
            l_color = dictarr[ str_name_data_annotation ];
            marker_colormap = { showscale: true, colorscale: "YlOrRd", reversescale: true }; // setting for showing colormap
            str_hovertemplate_additional_line = "<b>" + str_name_option_selected + "</b>: <b>%{marker.color:." + ( ( str_name_option_selected == 'MT Percentage' ) ? 1 : 0 ).toString( ) + "f}</b><br>"; // only MT percentage data will be shown as floating point numbers
            break;
        }
        // update x-axis values from gene expression data
        if ( object_data.main.scatter.x_axis_col_name == '__x_axis_gene_expr__' ) {
            let gene_symbol_x_axis = object_data.main.scatter.x_axis_option,
                l_expr_x_axis = [ ];
            await UpdateCache( object_cache, preprocess_expr, [ gene_symbol_x_axis ] ); // retrieve the single cells expression data of the current gene.
            l_expr_x_axis = ArrayBooleanIndexing( object_data.main.gene.expr.data[ gene_symbol_x_axis ].data, object_data.main.umap.l_bool_index_current_cells ); // retrieve x-axis values
            dictarr[ '__x_axis_gene_expr__' ] = l_expr_x_axis;
        } else {
            object_data.main.scatter.x_axis_col_name = ( object_data.main.umap.flag_combined_tissues ) ? 'UMAP_1' : 'tissue_UMAP_1';
        }
        if ( object_data.main.scatter.y_axis_col_name == '__y_axis_gene_expr__' ) {
            let gene_symbol_y_axis = object_data.main.scatter.y_axis_option,
                l_expr_y_axis = [ ];
            await UpdateCache( object_cache, preprocess_expr, [ gene_symbol_y_axis ] ); // retrieve the single cells expression data of the current gene.
            l_expr_y_axis = ArrayBooleanIndexing( object_data.main.gene.expr.data[ gene_symbol_y_axis ].data, object_data.main.umap.l_bool_index_current_cells ); // retrieve y-axis values
            dictarr[ '__y_axis_gene_expr__' ] = l_expr_y_axis;
        }  else {
            object_data.main.scatter.y_axis_col_name = ( object_data.main.umap.flag_combined_tissues ) ? 'UMAP_2' : 'tissue_UMAP_2';
        }
        // complete the selection process by re-enabling the selection button
        $id( 'select_scatter_x_axis' ).disabled = false; 
        $id( 'select_scatter_y_axis' ).disabled = false; 
        object_data.setting.plot.single_cell_scatter.int_factor_axis_range_to_marker_size = object_data.main.scatter.flag_axis_using_gene_expr ? 6 : 30;
        object_data.setting.plot.single_cell_scatter.marker_size = ResponsiveMarkerSize( plotSingleCellAnnotation, object_data.setting.plot.single_cell_scatter.int_factor_axis_range_to_marker_size, object_data.setting.plot.single_cell_scatter.int_factor_marker_size_multiply );
        if ( flag_restyle ) { // restyle existing plot
            console.log( "[PlotCellAnnotation] restyling existing plot for", str_name_data_annotation );
            let update = {
                hovertemplate: [ str_hovertemplate_additional_line + "tissue: <b>%{customdata[1]}</b>",
                                "cell type: <b>%{customdata[0]}</b>",
                                "",
                                object_data.main.scatter.x_label + ": <b>%{x:.2f}</b>",
                                object_data.main.scatter.y_label + ": <b>%{y:.2f}</b>",
                                "%{customdata[2]}",
                                "<extra></extra>" ].join( "<br>" ),
                marker: { 
                    color: l_color,
                    size: object_data.setting.plot.single_cell_scatter.marker_size
                } };
            update.marker = { ... update.marker, ... marker_colormap }; // update setting for showing colormap
            Plotly.restyle( plotSingleCellAnnotation, update, [ 0 ] );
        } 
        else {
            let trace = {
                mode: 'markers',
                type: 'scattergl',
                x: dictarr[ object_data.main.scatter.x_axis_col_name ],
                y: dictarr[ object_data.main.scatter.y_axis_col_name ],
                customdata: ArrayTranspose( [ dictarr[ col_name_cell_type ],
                                              dictarr[ "tissue" ],
                                              dictarr[ 'id_cell' ]
                                            ] ),
                hovertemplate: [ str_hovertemplate_additional_line + "tissue: <b>%{customdata[1]}</b>",
                                "cell type: <b>%{customdata[0]}</b>",
                                "",
                                object_data.main.scatter.x_label + ": <b>%{x:.2f}</b>",
                                object_data.main.scatter.y_label + ": <b>%{y:.2f}</b>",
                                "%{customdata[2]}",
                                "<extra></extra>" ].join( "<br>" ),
                marker: { 
                    color: l_color,
                    size: object_data.setting.plot.single_cell_scatter.marker_size            
                },
                line: { 
                    width: 0
                },
            };
            trace.marker = { ... trace.marker, ... marker_colormap }; // update setting for showing colormap
            Plotly.newPlot( plotSingleCellAnnotation, [ trace ], { margin: { l: 65, r: 65, b: 90, t: 50, pad: 4 }, xaxis: { title : object_data.main.scatter.x_label, zeroline: false, showgrid : false }, yaxis: { title : object_data.main.scatter.y_label, zeroline: false, showgrid : false }, hovermode: "closest" }, { ... config_plotly, filename: name_application + '__Annotation', height: 1100, width: 1100 } ); // draw scatter plot of the annotation
            
            // add interactions
            plotSingleCellAnnotation.on( 'plotly_click', async function( data ) {
                PlotlyCallBackSingleCellMarker( data.points[ 0 ] );
            });
            plotSingleCellAnnotation.on( "plotly_relayout" , function( ed ) {
                PlotlySynchronizedLayout( ed, l_div_plotly );
            });
        }
    }
    
    async function PlotlyCallBackSingleCellMarker( clicked_point ) { // load the cell-type marker gene table of the cell-type to which the clicked cell belongs
        console.log( clicked_point );
        let index_point = clicked_point.pointIndex, dictarr = object_data.main.umap.dictarr;
        let str_tissue = dictarr[ "tissue" ][ index_point ], str_tissue_cell_type = dictarr[ "tissue_cell_type" ][ index_point ], str_key = str_tissue + '|' + str_tissue_cell_type;
        console.log( str_key );
        let object_cache = object_data.main.cell_marker;
        await UpdateCache( object_cache, null, [ str_key ] ); // update cache of cell marker data
        Display_DataTable( 'container_datatable_cell_marker', 'datatable_cell_marker', object_cache.data[ str_key ].data, [ 'p_val_adj', 'gene', 'avg_log2FC', 'pct.1', 'pct.2' ] );
        // change panel name
        document.getElementById( "panel_name_cellmarker").innerHTML = "Cell Marker Table for " + str_tissue_cell_type + ' (' + str_tissue + ')';
        // change datatable name 
        object_data.main.datatable[ 'datatable_cell_marker' ].name = `cell_marker.${str_tissue}.${str_tissue_cell_type}`;
        // show datatable panel
        $( "#container_panel_datatable_cell_marker" ).collapse( 'show' )
    }
    
    function ResponsiveMarkerSize( div, int_factor_axis_range_to_marker_size = 6, int_factor_marker_size_multiply = 1 ) { // get marker size according to the given parameter and x- and y-axis range of the division containing a plotly plot
        if ( ! ( 'layout' in div ) ) {
            return object_data.setting.plot.single_cell_scatter.marker_size; // if plotly has not been drawn, use the default value for the marker size
        }
        let x = div.layout.xaxis, y = div.layout.yaxis,
            int_marker_size_to_be_updated = Math.max( 2, Math.min( 15, Math.min( Math.ceil( int_factor_axis_range_to_marker_size / ( x.range[ 1 ] - x.range[ 0 ] ) ), Math.ceil( int_factor_axis_range_to_marker_size / ( y.range[ 1 ] - y.range[ 0 ] ) ) ) * int_factor_marker_size_multiply ) ); // max value is 15, min value is 2
        return int_marker_size_to_be_updated;
    }
    // synchronization of zoom levels across plots
    function PlotlySynchronizedLayout( ed, divs ) {
        if ( Object.entries( ed ).length === 0 ) { return; }
        // broadcasting updated zoom level across plots
        divs.forEach( ( div, i ) => {
            let x = div.layout.xaxis, y = div.layout.yaxis, flag_relayout = false, update = { }; // a flag indicating whether a relayout event is required
            // ignore when layouts are changed due to autoranging
            let int_marker_size_current = div.data[ 0 ].marker.size;
            int_marker_size_to_be_updated = ResponsiveMarkerSize( div, object_data.setting.plot.single_cell_scatter.int_factor_axis_range_to_marker_size, object_data.setting.plot.single_cell_scatter.int_factor_marker_size_multiply );
            flag_restyle = int_marker_size_to_be_updated != int_marker_size_current;

            if ( flag_restyle ) { // restyle existing plot
                console.log( "[UpdateGene] restyling marker size to ", int_marker_size_to_be_updated );
                let update = { marker : div.data[ 0 ].marker };
                update.marker.size = int_marker_size_to_be_updated;
                object_data.setting.plot.single_cell_scatter.marker_size = int_marker_size_to_be_updated; // update marker size
                Plotly.restyle( div, update, [ 0 ] );
            } 
            
            if ( ( ed[ "xaxis.autorange" ] && x.autorange ) && ( ed[ "yaxis.autorange" ] && y.autorange ) ) { 
                return;
            }
            // update x-axis
            if ( x.range[ 0 ] != ed[ "xaxis.range[0]" ] || x.range[ 1 ] != ed[ "xaxis.range[1]" ] ) {
                flag_relayout = true;
                update = { ... update, 'xaxis.range[0]': ed[ "xaxis.range[0]" ], 'xaxis.range[1]': ed[ "xaxis.range[1]" ], 'xaxis.autorange': ed[ "xaxis.autorange" ] };
            }
            // update y-axis
            if ( y.range[ 0 ] != ed[ "yaxis.range[0]" ] || y.range[ 1 ] != ed[ "yaxis.range[1]" ] ) {
                flag_relayout = true;
                update = { ... update, 'yaxis.range[0]': ed[ "yaxis.range[0]" ], 'yaxis.range[1]': ed[ "yaxis.range[1]" ], 'yaxis.autorange': ed[ "yaxis.autorange" ] };
            }
            if ( flag_relayout ) {
                Plotly.relayout( div, update ); // perform relayout operation
            }
            
        });
    }
    var l_div_plotly = [ plotSingleCellAnnotation, plotSingleCellExpression ];

    function Display_DataTable( id_datatable_container, id_datatable, dictarr, l_col_selected = null ) { // load datatable with loaded datasets
        let object_datatable = object_data.main.datatable[ id_datatable ], columns = [ ], dataSet = [ ], df = null, l_col = [ ];

        df = new DataFrame( dictarr );
        object_datatable.df = df; // save the reference to the dataframe
        if ( ! ( l_col_selected === null ) ) {
            df = df.select( ... l_col_selected ); // select appropriate columns before visualization
        }
        dataSet = df.toArray( ), l_col = df.listColumns( );
        for ( let i = 0; i < l_col.length; i ++ ) { columns.push( { title: l_col[ i ] } ); } // set columns for DataTable.js
        $( '#' + id_datatable_container ).empty( ); // empty the datatable and redraw the datatable
        $( '#' + id_datatable_container ).append( '<table id="' + id_datatable + '" class="display" width="100%"></table>' );
        $( '#' + id_datatable ).DataTable( { data: dataSet, columns: columns } ); // display the datatable
    }
    function Download_DataTable( id_datatable ) { // download analysis result of vaccine candidate as a csv file
        let string_csv_datatable = '', object_datatable = object_data.main.datatable[ id_datatable ], df = object_datatable.df;
        if ( df != null ) {
            string_csv_datatable = df.toCSV( true );
        }
        let blob = new Blob( [ string_csv_datatable ], { type: "text/plain;charset=utf-8" } );
        name_file = object_datatable.name + '.datatable.csv' // change the extension of the filename
        saveAs( blob, name_file ); // save file 
    }

    function SortBoxTrace( trace_a, trace_b ) { return - ( trace_a.q3 - trace_b.q3 ) } // Sort Plotly Boxplot traces based on parameter values
    function PreviewGene( gene_symbol ) { // draw preview of expression of the current gene using boxplots
        
        let window_width = window.innerWidth,
            window_height = window.innerHeight,
            window_width_excluding_sidebar = $id( "controlBox_plotly_graph_main" ).offsetWidth,
            preview_height = parseInt( Math.min( Math.max( 175, window_height / 3.5 ), 500 ) ), // max height is 500px
            preview_width = window_width_excluding_sidebar,
            div_preview = $id( "plotly_searchbox_main_preview" );
        div_preview.style.height = ( preview_height ).toString( ) + 'px';
        div_preview.style.width = ( preview_width ).toString( ) + 'px';
        
        let index_gene = object_data.main.gene.expr.dict_key_to_index[ gene_symbol ][ 0 ],
            l_trace = [ ], dictdictarr = object_data.main.gene_summary.boxplot.dictdictarr;
        for ( let category in dictdictarr ) {
            let trace = { };
            trace.lowerfence = [ dictdictarr[ category ][ 'MIN' ][ index_gene ] / 1000 ]; // scale back to original expression scale (the data became integer during delivery to reduce its size)
            trace.q1 = [ dictdictarr[ category ][ 'Q1' ][ index_gene ] / 1000 ];
            trace.median = [ dictdictarr[ category ][ 'Q2' ][ index_gene ] / 1000 ];
            trace.q3 = [ dictdictarr[ category ][ 'Q3' ][ index_gene ] / 1000 ];
            trace.upperfence = [ dictdictarr[ category ][ 'MAX' ][ index_gene ] / 1000 ];
            trace.name = category;
            trace.line = { color: object_data.setting.plot.preview_gene.dict_category_to_color[ category ] }; // use pre-defined color for each tissue
            trace.type = 'box';
            trace.x0 = category;
            l_trace.push( trace );
        }
        l_trace = l_trace.sort( SortBoxTrace );
//         document.getElementById( "plotly_searchbox_main_preview" ).style.height = "350px";
        Plotly.newPlot( "plotly_searchbox_main_preview", l_trace, { margin: { l: 65, r: 65, b: 90, t: 50, pad: 4 }, yaxis: { title : "Relative Expression of " + gene_symbol } }, { filename: name_application + '__Preview_Expression_of_' + gene_symbol, height: preview_height, width: preview_width, ... config_plotly } ); // draw boxplot of expression of the gene 
    }
    function Toggle_Sidebar( ) {
        // toggle side menu
        let id_container = "#container_main_sidebar"
        if( $( id_container ).hasClass( 'toggled' ) ) { // when sidebar is shown
            $( id_container ).removeClass( 'toggled' ); // side sidebar
            button_open_sidebar.style.display = "block"; // show the button opening sidebar
        } else { // when sidebar is hidden
            $( id_container ).addClass( 'toggled' ); // show the sidebar
            button_open_sidebar.style.display = "none"; // hide the button opening sidebar
        }
    }
    $( "#container_main_sidebar" ).on( "toggle.sidebar.complete", function ( event ) {
        window.dispatchEvent( new Event( 'resize' ) ); // trigger resize events to allow Plotly plots to resize its plot
    } );
    $( "#container_main_sidebar" ).on( "pin.sidebar.complete", function ( event ) {
        window.dispatchEvent( new Event( 'resize' ) ); // trigger resize events to allow Plotly plots to resize its plot
    } );
    $( "#container_main_sidebar" ).on( "pin.sidebar.hidden", function ( event ) {
         button_open_sidebar.style.display = "block";
    } );
    // Get the button_open_sidebar:
    button_open_sidebar = document.getElementById( "button_open_sidebar" );


</script>
</body>
</html> 